// @turnkey/sdk-types/dist/index.mjs
var SessionType;
(function (SessionType) {
    SessionType["READ_ONLY"] = "SESSION_TYPE_READ_ONLY";
    SessionType["READ_WRITE"] = "SESSION_TYPE_READ_WRITE";
})(SessionType || (SessionType = {}));
var AuthAction;
(function (AuthAction) {
    AuthAction["LOGIN"] = "LOGIN";
    AuthAction["SIGNUP"] = "SIGNUP";
})(AuthAction || (AuthAction = {}));
var TurnkeyErrorCodes;
(function (TurnkeyErrorCodes) {
    TurnkeyErrorCodes["UNKNOWN"] = "UNKNOWN";
    TurnkeyErrorCodes["NETWORK_ERROR"] = "NETWORK_ERROR";
    TurnkeyErrorCodes["KEY_PAIR_CLEANUP_ERROR"] = "KEY_PAIR_CLEANUP_ERROR";
    TurnkeyErrorCodes["LOGOUT_ERROR"] = "LOGOUT_ERROR";
    TurnkeyErrorCodes["CREATE_PASSKEY_ERROR"] = "CREATE_PASSKEY_ERROR";
    TurnkeyErrorCodes["SELECT_PASSKEY_CANCELLED"] = "SELECT_PASSKEY_CANCELLED";
    TurnkeyErrorCodes["CONNECT_WALLET_CANCELLED"] = "CONNECT_WALLET_CANCELLED";
    TurnkeyErrorCodes["WALLET_CONNECT_INITIALIZATION_ERROR"] = "WALLET_CONNECT_INITIALIZATION_ERROR";
    TurnkeyErrorCodes["WALLET_CONNECT_EXPIRED"] = "WALLET_CONNECT_EXPIRED";
    TurnkeyErrorCodes["PASSKEY_SIGNUP_AUTH_ERROR"] = "PASSKEY_SIGNUP_AUTH_ERROR";
    TurnkeyErrorCodes["PASSKEY_LOGIN_AUTH_ERROR"] = "PASSKEY_LOGIN_AUTH_ERROR";
    TurnkeyErrorCodes["WALLET_BUILD_LOGIN_REQUEST_ERROR"] = "WALLET_BUILD_LOGIN_REQUEST_ERROR";
    TurnkeyErrorCodes["WALLET_LOGIN_AUTH_ERROR"] = "WALLET_LOGIN_AUTH_ERROR";
    TurnkeyErrorCodes["WALLET_SIGNUP_AUTH_ERROR"] = "WALLET_SIGNUP_AUTH_ERROR";
    TurnkeyErrorCodes["WALLET_LOGIN_OR_SIGNUP_ERROR"] = "WALLET_LOGIN_OR_SIGNUP_ERROR";
    TurnkeyErrorCodes["INIT_OTP_ERROR"] = "INIT_OTP_ERROR";
    TurnkeyErrorCodes["VERIFY_OTP_ERROR"] = "VERIFY_OTP_ERROR";
    TurnkeyErrorCodes["OTP_LOGIN_ERROR"] = "OTP_LOGIN_ERROR";
    TurnkeyErrorCodes["OTP_SIGNUP_ERROR"] = "OTP_SIGNUP_ERROR";
    TurnkeyErrorCodes["OTP_COMPLETION_ERROR"] = "OTP_COMPLETION_ERROR";
    TurnkeyErrorCodes["OAUTH_LOGIN_ERROR"] = "OAUTH_LOGIN_ERROR";
    TurnkeyErrorCodes["OAUTH_SIGNUP_ERROR"] = "OAUTH_SIGNUP_ERROR";
    TurnkeyErrorCodes["ACCOUNT_FETCH_ERROR"] = "ACCOUNT_FETCH_ERROR";
    TurnkeyErrorCodes["INVALID_OTP_CODE"] = "INVALID_OTP_CODE";
    TurnkeyErrorCodes["FETCH_WALLETS_ERROR"] = "FETCH_WALLETS_ERROR";
    TurnkeyErrorCodes["FETCH_WALLET_ACCOUNTS_ERROR"] = "FETCH_WALLET_ACCOUNTS_ERROR";
    TurnkeyErrorCodes["FETCH_PRIVATE_KEYS_ERROR"] = "FETCH_PRIVATE_KEYS_ERROR";
    TurnkeyErrorCodes["SIGN_MESSAGE_ERROR"] = "SIGN_MESSAGE_ERROR";
    TurnkeyErrorCodes["SIGN_TRANSACTION_ERROR"] = "SIGN_TRANSACTION_ERROR";
    TurnkeyErrorCodes["POLL_TRANSACTION_STATUS_ERROR"] = "POLL_TRANSACTION_STATUS_ERROR";
    TurnkeyErrorCodes["SIGN_AND_SEND_TRANSACTION_ERROR"] = "SIGN_AND_SEND_TRANSACTION_ERROR";
    TurnkeyErrorCodes["ETH_SEND_TRANSACTION_ERROR"] = "ETH_SEND_TRANSACTION_ERROR";
    TurnkeyErrorCodes["SOL_SEND_TRANSACTION_ERROR"] = "SOL_SEND_TRANSACTION_ERROR";
    TurnkeyErrorCodes["FETCH_USER_ERROR"] = "FETCH_USERS_ERROR";
    TurnkeyErrorCodes["CREATE_WALLET_ERROR"] = "CREATE_WALLET_ERROR";
    TurnkeyErrorCodes["CREATE_WALLET_ACCOUNT_ERROR"] = "CREATE_WALLET_ACCOUNT_ERROR";
    TurnkeyErrorCodes["EXPORT_WALLET_ERROR"] = "EXPORT_WALLET_ERROR";
    TurnkeyErrorCodes["EXPORT_PRIVATE_KEY_ERROR"] = "EXPORT_PRIVATE_KEY_ERROR";
    TurnkeyErrorCodes["EXPORT_WALLET_ACCOUNT_ERROR"] = "EXPORT_WALLET_ACCOUNT_ERROR";
    TurnkeyErrorCodes["IMPORT_WALLET_ERROR"] = "IMPORT_WALLET_ERROR";
    TurnkeyErrorCodes["DELETE_SUB_ORGANIZATION_ERROR"] = "DELETE_SUB_ORGANIZATION_ERROR";
    TurnkeyErrorCodes["CREATE_SUB_ORGANIZATION_ERROR"] = "CREATE_SUB_ORGANIZATION_ERROR";
    TurnkeyErrorCodes["CREATE_USERS_ERROR"] = "CREATE_USERS_ERROR";
    TurnkeyErrorCodes["FETCH_BOOT_PROOF_ERROR"] = "FETCH_BOOT_PROOF_ERROR";
    TurnkeyErrorCodes["VERIFY_APP_PROOFS_ERROR"] = "VERIFY_APP_PROOFS_ERROR";
    TurnkeyErrorCodes["CREATE_POLICY_ERROR"] = "CREATE_POLICY_ERROR";
    TurnkeyErrorCodes["STORE_SESSION_ERROR"] = "STORE_SESSION_ERROR";
    TurnkeyErrorCodes["CLEAR_SESSION_ERROR"] = "CLEAR_SESSION_ERROR";
    TurnkeyErrorCodes["CLEAR_ALL_SESSIONS_ERROR"] = "CLEAR_ALL_SESSIONS_ERROR";
    TurnkeyErrorCodes["REFRESH_SESSION_ERROR"] = "REFRESH_SESSION_ERROR";
    TurnkeyErrorCodes["GET_SESSION_ERROR"] = "GET_SESSION_ERROR";
    TurnkeyErrorCodes["GET_WALLET_PROVIDERS_ERROR"] = "GET_WALLET_PROVIDERS_ERROR";
    TurnkeyErrorCodes["GET_ALL_SESSIONS_ERROR"] = "GET_ALL_SESSIONS_ERROR";
    TurnkeyErrorCodes["SET_ACTIVE_SESSION_ERROR"] = "SET_ACTIVE_SESSION_ERROR";
    TurnkeyErrorCodes["GET_ACTIVE_SESSION_KEY_ERROR"] = "GET_ACTIVE_SESSION_KEY_ERROR";
    TurnkeyErrorCodes["CLEAR_UNUSED_KEY_PAIRS_ERROR"] = "CLEAR_UNUSED_KEY_PAIRS_ERROR";
    TurnkeyErrorCodes["CREATE_API_KEY_PAIR_ERROR"] = "CREATE_API_KEY_PAIR_ERROR";
    TurnkeyErrorCodes["GET_PROXY_AUTH_CONFIG_ERROR"] = "GET_PROXY_AUTH_CONFIG_ERROR";
    TurnkeyErrorCodes["UPDATE_USER_EMAIL_ERROR"] = "UPDATE_USER_EMAIL_ERROR";
    TurnkeyErrorCodes["UPDATE_USER_NAME_ERROR"] = "UPDATE_USER_NAME_ERROR";
    TurnkeyErrorCodes["UPDATE_USER_PHONE_NUMBER_ERROR"] = "UPDATE_USER_PHONE_NUMBER_ERROR";
    TurnkeyErrorCodes["ADD_OAUTH_PROVIDER_ERROR"] = "ADD_OAUTH_PROVIDER_ERROR";
    TurnkeyErrorCodes["REMOVE_OAUTH_PROVIDER_ERROR"] = "REMOVE_OAUTH_PROVIDER_ERROR";
    TurnkeyErrorCodes["ADD_PASSKEY_ERROR"] = "ADD_PASSKEY_ERROR";
    TurnkeyErrorCodes["REMOVE_PASSKEY_ERROR"] = "REMOVE_PASSKEY_ERROR";
    TurnkeyErrorCodes["CONNECT_WALLET_ACCOUNT_ERROR"] = "CONNECT_WALLET_ACCOUNT_ERROR";
    TurnkeyErrorCodes["DISCONNECT_WALLET_ACCOUNT_ERROR"] = "DISCONNECT_WALLET_ACCOUNT_ERROR";
    TurnkeyErrorCodes["SWITCH_WALLET_CHAIN_ERROR"] = "SWITCH_WALLET_CHAIN_ERROR";
    TurnkeyErrorCodes["ONRAMP_ERROR"] = "ONRAMP_ERROR";
    TurnkeyErrorCodes["MAX_OTP_INITIATED_ERROR"] = "MAX_OTP_INITIATED_ERROR";
    TurnkeyErrorCodes["CLIENT_NOT_INITIALIZED"] = "CLIENT_NOT_INITIALIZED";
    TurnkeyErrorCodes["WALLET_MANAGER_COMPONENT_NOT_INITIALIZED"] = "WALLET_MANAGER_COMPONENT_NOT_INITIALIZED";
    TurnkeyErrorCodes["CONFIG_NOT_INITIALIZED"] = "CONFIG_NOT_INITIALIZED";
    TurnkeyErrorCodes["AUTH_METHOD_NOT_ENABLED"] = "AUTH_METHOD_NOT_ENABLED";
    TurnkeyErrorCodes["FEATURE_NOT_ENABLED"] = "FEATURE_NOT_ENABLED";
    TurnkeyErrorCodes["INITIALIZE_CLIENT_ERROR"] = "INITIALIZE_CLIENT_ERROR";
    TurnkeyErrorCodes["INITIALIZE_SESSION_ERROR"] = "INITIALIZE_SESSION_ERROR";
    TurnkeyErrorCodes["SCHEDULE_SESSION_EXPIRY_ERROR"] = "SCHEDULE_SESSION_EXPIRY_ERROR";
    TurnkeyErrorCodes["HANDLE_POST_AUTH_ERROR"] = "HANDLE_POST_AUTH_ERROR";
    TurnkeyErrorCodes["HANDLE_POST_LOGOUT_ERROR"] = "HANDLE_POST_LOGOUT_ERROR";
    TurnkeyErrorCodes["CLEAR_SESSION_TIMEOUTS_ERROR"] = "CLEAR_SESSION_TIMEOUTS_ERROR";
    TurnkeyErrorCodes["UPDATE_USER_ERROR"] = "UPDATE_USER_ERROR";
    TurnkeyErrorCodes["ACCOUNT_ALREADY_EXISTS"] = "ACCOUNT_ALREADY_EXISTS";
    TurnkeyErrorCodes["INITIALIZE_IFRAME_ERROR"] = "INITIALIZE_IFRAME_ERROR";
    TurnkeyErrorCodes["PLATFORM_MISMATCH"] = "PLATFORM_MISMATCH";
    TurnkeyErrorCodes["UNSUPPORTED_PLATFORM"] = "UNSUPPORTED_PLATFORM";
    TurnkeyErrorCodes["USER_CANCELED"] = "USER_CANCELED";
    TurnkeyErrorCodes["BAD_RESPONSE"] = "BAD_RESPONSE";
    TurnkeyErrorCodes["OIDC_TOKEN_ERROR"] = "OIDC_TOKEN_ERROR";
    TurnkeyErrorCodes["MISSING_PARAMS"] = "MISSING_PARAMS";
    TurnkeyErrorCodes["INVALID_CONFIGURATION"] = "INVALID_CONFIGURATION";
    TurnkeyErrorCodes["INVALID_REQUEST"] = "INVALID_REQUEST";
    TurnkeyErrorCodes["VALIDATION_ERROR"] = "VALIDATION_ERROR";
    TurnkeyErrorCodes["SESSION_EXPIRED"] = "SESSION_EXPIRED";
    TurnkeyErrorCodes["NO_SESSION_FOUND"] = "NO_SESSION_FOUND";
    TurnkeyErrorCodes["NO_WALLET_FOUND"] = "NO_WALLET_FOUND";
    TurnkeyErrorCodes["NO_WALLETS_FOUND"] = "NO_WALLETS_FOUND";
    TurnkeyErrorCodes["NOT_FOUND"] = "NOT_FOUND";
    TurnkeyErrorCodes["INTERNAL_ERROR"] = "INTERNAL_ERROR";
    TurnkeyErrorCodes["UNAUTHORIZED"] = "UNAUTHORIZED";
    TurnkeyErrorCodes["FORBIDDEN"] = "FORBIDDEN";
    TurnkeyErrorCodes["BAD_REQUEST"] = "BAD_REQUEST";
    TurnkeyErrorCodes["CONFLICT"] = "CONFLICT";
    TurnkeyErrorCodes["TIMEOUT"] = "TIMEOUT";
    TurnkeyErrorCodes["SERVICE_UNAVAILABLE"] = "SERVICE_UNAVAILABLE";
    TurnkeyErrorCodes["GATEWAY_TIMEOUT"] = "GATEWAY_TIMEOUT";
})(TurnkeyErrorCodes || (TurnkeyErrorCodes = {}));
class TurnkeyError extends Error {
    constructor(message, code, cause) {
        super(message);
        this.code = code;
        this.cause = cause;
        this.name = "TurnkeyError";
    }
}
class TurnkeyNetworkError extends TurnkeyError {
    constructor(message, statusCode, code, cause) {
        super(message, code, cause);
        this.statusCode = statusCode;
        this.name = "TurnkeyNetworkError";
    }
}
var FiatOnRampProvider;
(function (FiatOnRampProvider) {
    FiatOnRampProvider["COINBASE"] = "FIAT_ON_RAMP_PROVIDER_COINBASE";
    FiatOnRampProvider["MOONPAY"] = "FIAT_ON_RAMP_PROVIDER_MOONPAY";
})(FiatOnRampProvider || (FiatOnRampProvider = {}));
var FiatOnRampCryptoCurrency;
(function (FiatOnRampCryptoCurrency) {
    FiatOnRampCryptoCurrency["BITCOIN"] = "FIAT_ON_RAMP_CRYPTO_CURRENCY_BTC";
    FiatOnRampCryptoCurrency["ETHEREUM"] = "FIAT_ON_RAMP_CRYPTO_CURRENCY_ETH";
    FiatOnRampCryptoCurrency["SOLANA"] = "FIAT_ON_RAMP_CRYPTO_CURRENCY_SOL";
    FiatOnRampCryptoCurrency["USDC"] = "FIAT_ON_RAMP_CRYPTO_CURRENCY_USDC";
})(FiatOnRampCryptoCurrency || (FiatOnRampCryptoCurrency = {}));
var FiatOnRampCurrency;
(function (FiatOnRampCurrency) {
    FiatOnRampCurrency["AUD"] = "FIAT_ON_RAMP_CURRENCY_AUD";
    FiatOnRampCurrency["BGN"] = "FIAT_ON_RAMP_CURRENCY_BGN";
    FiatOnRampCurrency["BRL"] = "FIAT_ON_RAMP_CURRENCY_BRL";
    FiatOnRampCurrency["CAD"] = "FIAT_ON_RAMP_CURRENCY_CAD";
    FiatOnRampCurrency["CHF"] = "FIAT_ON_RAMP_CURRENCY_CHF";
    FiatOnRampCurrency["COP"] = "FIAT_ON_RAMP_CURRENCY_COP";
    FiatOnRampCurrency["CZK"] = "FIAT_ON_RAMP_CURRENCY_CZK";
    FiatOnRampCurrency["DKK"] = "FIAT_ON_RAMP_CURRENCY_DKK";
    FiatOnRampCurrency["DOP"] = "FIAT_ON_RAMP_CURRENCY_DOP";
    FiatOnRampCurrency["EGP"] = "FIAT_ON_RAMP_CURRENCY_EGP";
    FiatOnRampCurrency["EUR"] = "FIAT_ON_RAMP_CURRENCY_EUR";
    FiatOnRampCurrency["GBP"] = "FIAT_ON_RAMP_CURRENCY_GBP";
    FiatOnRampCurrency["HKD"] = "FIAT_ON_RAMP_CURRENCY_HKD";
    FiatOnRampCurrency["IDR"] = "FIAT_ON_RAMP_CURRENCY_IDR";
    FiatOnRampCurrency["ILS"] = "FIAT_ON_RAMP_CURRENCY_ILS";
    FiatOnRampCurrency["JOD"] = "FIAT_ON_RAMP_CURRENCY_JOD";
    FiatOnRampCurrency["KES"] = "FIAT_ON_RAMP_CURRENCY_KES";
    FiatOnRampCurrency["KWD"] = "FIAT_ON_RAMP_CURRENCY_KWD";
    FiatOnRampCurrency["LKR"] = "FIAT_ON_RAMP_CURRENCY_LKR";
    FiatOnRampCurrency["MXN"] = "FIAT_ON_RAMP_CURRENCY_MXN";
    FiatOnRampCurrency["NGN"] = "FIAT_ON_RAMP_CURRENCY_NGN";
    FiatOnRampCurrency["NOK"] = "FIAT_ON_RAMP_CURRENCY_NOK";
    FiatOnRampCurrency["NZD"] = "FIAT_ON_RAMP_CURRENCY_NZD";
    FiatOnRampCurrency["OMR"] = "FIAT_ON_RAMP_CURRENCY_OMR";
    FiatOnRampCurrency["PEN"] = "FIAT_ON_RAMP_CURRENCY_PEN";
    FiatOnRampCurrency["PLN"] = "FIAT_ON_RAMP_CURRENCY_PLN";
    FiatOnRampCurrency["RON"] = "FIAT_ON_RAMP_CURRENCY_RON";
    FiatOnRampCurrency["SEK"] = "FIAT_ON_RAMP_CURRENCY_SEK";
    FiatOnRampCurrency["THB"] = "FIAT_ON_RAMP_CURRENCY_THB";
    FiatOnRampCurrency["TRY"] = "FIAT_ON_RAMP_CURRENCY_TRY";
    FiatOnRampCurrency["TWD"] = "FIAT_ON_RAMP_CURRENCY_TWD";
    FiatOnRampCurrency["USD"] = "FIAT_ON_RAMP_CURRENCY_USD";
    FiatOnRampCurrency["VND"] = "FIAT_ON_RAMP_CURRENCY_VND";
    FiatOnRampCurrency["ZAR"] = "FIAT_ON_RAMP_CURRENCY_ZAR";
})(FiatOnRampCurrency || (FiatOnRampCurrency = {}));
var FiatOnRampBlockchainNetwork;
(function (FiatOnRampBlockchainNetwork) {
    FiatOnRampBlockchainNetwork["BITCOIN"] = "FIAT_ON_RAMP_BLOCKCHAIN_NETWORK_BITCOIN";
    FiatOnRampBlockchainNetwork["ETHEREUM"] = "FIAT_ON_RAMP_BLOCKCHAIN_NETWORK_ETHEREUM";
    FiatOnRampBlockchainNetwork["SOLANA"] = "FIAT_ON_RAMP_BLOCKCHAIN_NETWORK_SOLANA";
    FiatOnRampBlockchainNetwork["BASE"] = "FIAT_ON_RAMP_BLOCKCHAIN_NETWORK_BASE";
})(FiatOnRampBlockchainNetwork || (FiatOnRampBlockchainNetwork = {}));
var FiatOnRampPaymentMethod;
(function (FiatOnRampPaymentMethod) {
    // Shared methods (supported by both MoonPay and Coinbase)
    FiatOnRampPaymentMethod["CREDIT_DEBIT_CARD"] = "FIAT_ON_RAMP_PAYMENT_METHOD_CREDIT_DEBIT_CARD";
    FiatOnRampPaymentMethod["APPLE_PAY"] = "FIAT_ON_RAMP_PAYMENT_METHOD_APPLE_PAY";
    // MoonPay-specific methods
    FiatOnRampPaymentMethod["GBP_BANK_TRANSFER"] = "FIAT_ON_RAMP_PAYMENT_METHOD_GBP_BANK_TRANSFER";
    FiatOnRampPaymentMethod["GBP_OPEN_BANKING_PAYMENT"] = "FIAT_ON_RAMP_PAYMENT_METHOD_GBP_OPEN_BANKING_PAYMENT";
    FiatOnRampPaymentMethod["GOOGLE_PAY"] = "FIAT_ON_RAMP_PAYMENT_METHOD_GOOGLE_PAY";
    FiatOnRampPaymentMethod["SEPA_BANK_TRANSFER"] = "FIAT_ON_RAMP_PAYMENT_METHOD_SEPA_BANK_TRANSFER";
    FiatOnRampPaymentMethod["PIX_INSTANT_PAYMENT"] = "FIAT_ON_RAMP_PAYMENT_METHOD_PIX_INSTANT_PAYMENT";
    FiatOnRampPaymentMethod["PAYPAL"] = "FIAT_ON_RAMP_PAYMENT_METHOD_PAYPAL";
    FiatOnRampPaymentMethod["VENMO"] = "FIAT_ON_RAMP_PAYMENT_METHOD_VENMO";
    FiatOnRampPaymentMethod["MOONPAY_BALANCE"] = "FIAT_ON_RAMP_PAYMENT_METHOD_MOONPAY_BALANCE";
    // Coinbase-specific methods
    FiatOnRampPaymentMethod["CRYPTO_ACCOUNT"] = "FIAT_ON_RAMP_PAYMENT_METHOD_CRYPTO_ACCOUNT";
    FiatOnRampPaymentMethod["FIAT_WALLET"] = "FIAT_ON_RAMP_PAYMENT_METHOD_FIAT_WALLET";
    FiatOnRampPaymentMethod["ACH_BANK_ACCOUNT"] = "FIAT_ON_RAMP_PAYMENT_METHOD_ACH_BANK_ACCOUNT";
})(FiatOnRampPaymentMethod || (FiatOnRampPaymentMethod = {}));
var OAuthProviders;
(function (OAuthProviders) {
    OAuthProviders["DISCORD"] = "discord";
    OAuthProviders["APPLE"] = "apple";
    OAuthProviders["GOOGLE"] = "google";
    OAuthProviders["FACEBOOK"] = "facebook";
    OAuthProviders["X"] = "x";
})(OAuthProviders || (OAuthProviders = {}));

export { AuthAction, FiatOnRampBlockchainNetwork, FiatOnRampCryptoCurrency, FiatOnRampCurrency, FiatOnRampPaymentMethod, FiatOnRampProvider, OAuthProviders, SessionType, TurnkeyError, TurnkeyErrorCodes, TurnkeyNetworkError };
