// @turnkey/core/dist/__generated__/sdk-client-base.mjs
import { StamperType } from '../__types__/enums.mjs';
import { TurnkeyRequestError } from '../__types__/error.mjs';
import { TERMINAL_ACTIVITY_STATUSES } from '../__types__/http.mjs';
import { VERSION } from './version.mjs';
import { TurnkeyError, TurnkeyErrorCodes } from '@turnkey/sdk-types';

/* @generated by codegen. DO NOT EDIT BY HAND */
class TurnkeySDKClientBase {
    constructor(config) {
        this.getActivity = async (input, stampWith) => {
            const session = await this.storageManager?.getActiveSession();
            return this.request("/public/v1/query/get_activity", {
                ...input,
                organizationId: input.organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
            }, stampWith);
        };
        this.stampGetActivity = async (input, stampWith) => {
            const activeStamper = this.getStamper(stampWith);
            if (!activeStamper) {
                return undefined;
            }
            const session = await this.storageManager?.getActiveSession();
            const fullUrl = this.config.apiBaseUrl + "/public/v1/query/get_activity";
            const body = {
                ...input,
                organizationId: input.organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
            };
            const stringifiedBody = JSON.stringify(body);
            const stamp = await activeStamper.stamp(stringifiedBody);
            return {
                body: stringifiedBody,
                stamp: stamp,
                url: fullUrl,
            };
        };
        this.getApiKey = async (input, stampWith) => {
            const session = await this.storageManager?.getActiveSession();
            return this.request("/public/v1/query/get_api_key", {
                ...input,
                organizationId: input.organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
            }, stampWith);
        };
        this.stampGetApiKey = async (input, stampWith) => {
            const activeStamper = this.getStamper(stampWith);
            if (!activeStamper) {
                return undefined;
            }
            const session = await this.storageManager?.getActiveSession();
            const fullUrl = this.config.apiBaseUrl + "/public/v1/query/get_api_key";
            const body = {
                ...input,
                organizationId: input.organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
            };
            const stringifiedBody = JSON.stringify(body);
            const stamp = await activeStamper.stamp(stringifiedBody);
            return {
                body: stringifiedBody,
                stamp: stamp,
                url: fullUrl,
            };
        };
        this.getApiKeys = async (input = {}, stampWith) => {
            const session = await this.storageManager?.getActiveSession();
            return this.request("/public/v1/query/get_api_keys", {
                ...input,
                organizationId: input.organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
            }, stampWith);
        };
        this.stampGetApiKeys = async (input, stampWith) => {
            const activeStamper = this.getStamper(stampWith);
            if (!activeStamper) {
                return undefined;
            }
            const session = await this.storageManager?.getActiveSession();
            const fullUrl = this.config.apiBaseUrl + "/public/v1/query/get_api_keys";
            const body = {
                ...input,
                organizationId: input.organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
            };
            const stringifiedBody = JSON.stringify(body);
            const stamp = await activeStamper.stamp(stringifiedBody);
            return {
                body: stringifiedBody,
                stamp: stamp,
                url: fullUrl,
            };
        };
        this.getAuthenticator = async (input, stampWith) => {
            const session = await this.storageManager?.getActiveSession();
            return this.request("/public/v1/query/get_authenticator", {
                ...input,
                organizationId: input.organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
            }, stampWith);
        };
        this.stampGetAuthenticator = async (input, stampWith) => {
            const activeStamper = this.getStamper(stampWith);
            if (!activeStamper) {
                return undefined;
            }
            const session = await this.storageManager?.getActiveSession();
            const fullUrl = this.config.apiBaseUrl + "/public/v1/query/get_authenticator";
            const body = {
                ...input,
                organizationId: input.organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
            };
            const stringifiedBody = JSON.stringify(body);
            const stamp = await activeStamper.stamp(stringifiedBody);
            return {
                body: stringifiedBody,
                stamp: stamp,
                url: fullUrl,
            };
        };
        this.getAuthenticators = async (input, stampWith) => {
            const session = await this.storageManager?.getActiveSession();
            return this.request("/public/v1/query/get_authenticators", {
                ...input,
                organizationId: input.organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
            }, stampWith);
        };
        this.stampGetAuthenticators = async (input, stampWith) => {
            const activeStamper = this.getStamper(stampWith);
            if (!activeStamper) {
                return undefined;
            }
            const session = await this.storageManager?.getActiveSession();
            const fullUrl = this.config.apiBaseUrl + "/public/v1/query/get_authenticators";
            const body = {
                ...input,
                organizationId: input.organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
            };
            const stringifiedBody = JSON.stringify(body);
            const stamp = await activeStamper.stamp(stringifiedBody);
            return {
                body: stringifiedBody,
                stamp: stamp,
                url: fullUrl,
            };
        };
        this.getBootProof = async (input, stampWith) => {
            const session = await this.storageManager?.getActiveSession();
            return this.request("/public/v1/query/get_boot_proof", {
                ...input,
                organizationId: input.organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
            }, stampWith);
        };
        this.stampGetBootProof = async (input, stampWith) => {
            const activeStamper = this.getStamper(stampWith);
            if (!activeStamper) {
                return undefined;
            }
            const session = await this.storageManager?.getActiveSession();
            const fullUrl = this.config.apiBaseUrl + "/public/v1/query/get_boot_proof";
            const body = {
                ...input,
                organizationId: input.organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
            };
            const stringifiedBody = JSON.stringify(body);
            const stamp = await activeStamper.stamp(stringifiedBody);
            return {
                body: stringifiedBody,
                stamp: stamp,
                url: fullUrl,
            };
        };
        this.getGasUsage = async (input, stampWith) => {
            const session = await this.storageManager?.getActiveSession();
            return this.request("/public/v1/query/get_gas_usage", {
                ...input,
                organizationId: input.organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
            }, stampWith);
        };
        this.stampGetGasUsage = async (input, stampWith) => {
            const activeStamper = this.getStamper(stampWith);
            if (!activeStamper) {
                return undefined;
            }
            const session = await this.storageManager?.getActiveSession();
            const fullUrl = this.config.apiBaseUrl + "/public/v1/query/get_gas_usage";
            const body = {
                ...input,
                organizationId: input.organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
            };
            const stringifiedBody = JSON.stringify(body);
            const stamp = await activeStamper.stamp(stringifiedBody);
            return {
                body: stringifiedBody,
                stamp: stamp,
                url: fullUrl,
            };
        };
        this.getLatestBootProof = async (input, stampWith) => {
            const session = await this.storageManager?.getActiveSession();
            return this.request("/public/v1/query/get_latest_boot_proof", {
                ...input,
                organizationId: input.organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
            }, stampWith);
        };
        this.stampGetLatestBootProof = async (input, stampWith) => {
            const activeStamper = this.getStamper(stampWith);
            if (!activeStamper) {
                return undefined;
            }
            const session = await this.storageManager?.getActiveSession();
            const fullUrl = this.config.apiBaseUrl + "/public/v1/query/get_latest_boot_proof";
            const body = {
                ...input,
                organizationId: input.organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
            };
            const stringifiedBody = JSON.stringify(body);
            const stamp = await activeStamper.stamp(stringifiedBody);
            return {
                body: stringifiedBody,
                stamp: stamp,
                url: fullUrl,
            };
        };
        this.getNonces = async (input, stampWith) => {
            const session = await this.storageManager?.getActiveSession();
            return this.request("/public/v1/query/get_nonces", {
                ...input,
                organizationId: input.organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
            }, stampWith);
        };
        this.stampGetNonces = async (input, stampWith) => {
            const activeStamper = this.getStamper(stampWith);
            if (!activeStamper) {
                return undefined;
            }
            const session = await this.storageManager?.getActiveSession();
            const fullUrl = this.config.apiBaseUrl + "/public/v1/query/get_nonces";
            const body = {
                ...input,
                organizationId: input.organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
            };
            const stringifiedBody = JSON.stringify(body);
            const stamp = await activeStamper.stamp(stringifiedBody);
            return {
                body: stringifiedBody,
                stamp: stamp,
                url: fullUrl,
            };
        };
        this.getOauth2Credential = async (input, stampWith) => {
            const session = await this.storageManager?.getActiveSession();
            return this.request("/public/v1/query/get_oauth2_credential", {
                ...input,
                organizationId: input.organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
            }, stampWith);
        };
        this.stampGetOauth2Credential = async (input, stampWith) => {
            const activeStamper = this.getStamper(stampWith);
            if (!activeStamper) {
                return undefined;
            }
            const session = await this.storageManager?.getActiveSession();
            const fullUrl = this.config.apiBaseUrl + "/public/v1/query/get_oauth2_credential";
            const body = {
                ...input,
                organizationId: input.organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
            };
            const stringifiedBody = JSON.stringify(body);
            const stamp = await activeStamper.stamp(stringifiedBody);
            return {
                body: stringifiedBody,
                stamp: stamp,
                url: fullUrl,
            };
        };
        this.getOauthProviders = async (input, stampWith) => {
            const session = await this.storageManager?.getActiveSession();
            return this.request("/public/v1/query/get_oauth_providers", {
                ...input,
                organizationId: input.organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
            }, stampWith);
        };
        this.stampGetOauthProviders = async (input, stampWith) => {
            const activeStamper = this.getStamper(stampWith);
            if (!activeStamper) {
                return undefined;
            }
            const session = await this.storageManager?.getActiveSession();
            const fullUrl = this.config.apiBaseUrl + "/public/v1/query/get_oauth_providers";
            const body = {
                ...input,
                organizationId: input.organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
            };
            const stringifiedBody = JSON.stringify(body);
            const stamp = await activeStamper.stamp(stringifiedBody);
            return {
                body: stringifiedBody,
                stamp: stamp,
                url: fullUrl,
            };
        };
        this.getOnRampTransactionStatus = async (input, stampWith) => {
            const session = await this.storageManager?.getActiveSession();
            return this.request("/public/v1/query/get_onramp_transaction_status", {
                ...input,
                organizationId: input.organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
            }, stampWith);
        };
        this.stampGetOnRampTransactionStatus = async (input, stampWith) => {
            const activeStamper = this.getStamper(stampWith);
            if (!activeStamper) {
                return undefined;
            }
            const session = await this.storageManager?.getActiveSession();
            const fullUrl = this.config.apiBaseUrl + "/public/v1/query/get_onramp_transaction_status";
            const body = {
                ...input,
                organizationId: input.organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
            };
            const stringifiedBody = JSON.stringify(body);
            const stamp = await activeStamper.stamp(stringifiedBody);
            return {
                body: stringifiedBody,
                stamp: stamp,
                url: fullUrl,
            };
        };
        this.getOrganizationConfigs = async (input, stampWith) => {
            const session = await this.storageManager?.getActiveSession();
            return this.request("/public/v1/query/get_organization_configs", {
                ...input,
                organizationId: input.organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
            }, stampWith);
        };
        this.stampGetOrganizationConfigs = async (input, stampWith) => {
            const activeStamper = this.getStamper(stampWith);
            if (!activeStamper) {
                return undefined;
            }
            const session = await this.storageManager?.getActiveSession();
            const fullUrl = this.config.apiBaseUrl + "/public/v1/query/get_organization_configs";
            const body = {
                ...input,
                organizationId: input.organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
            };
            const stringifiedBody = JSON.stringify(body);
            const stamp = await activeStamper.stamp(stringifiedBody);
            return {
                body: stringifiedBody,
                stamp: stamp,
                url: fullUrl,
            };
        };
        this.getPolicy = async (input, stampWith) => {
            const session = await this.storageManager?.getActiveSession();
            return this.request("/public/v1/query/get_policy", {
                ...input,
                organizationId: input.organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
            }, stampWith);
        };
        this.stampGetPolicy = async (input, stampWith) => {
            const activeStamper = this.getStamper(stampWith);
            if (!activeStamper) {
                return undefined;
            }
            const session = await this.storageManager?.getActiveSession();
            const fullUrl = this.config.apiBaseUrl + "/public/v1/query/get_policy";
            const body = {
                ...input,
                organizationId: input.organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
            };
            const stringifiedBody = JSON.stringify(body);
            const stamp = await activeStamper.stamp(stringifiedBody);
            return {
                body: stringifiedBody,
                stamp: stamp,
                url: fullUrl,
            };
        };
        this.getPolicyEvaluations = async (input, stampWith) => {
            const session = await this.storageManager?.getActiveSession();
            return this.request("/public/v1/query/get_policy_evaluations", {
                ...input,
                organizationId: input.organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
            }, stampWith);
        };
        this.stampGetPolicyEvaluations = async (input, stampWith) => {
            const activeStamper = this.getStamper(stampWith);
            if (!activeStamper) {
                return undefined;
            }
            const session = await this.storageManager?.getActiveSession();
            const fullUrl = this.config.apiBaseUrl + "/public/v1/query/get_policy_evaluations";
            const body = {
                ...input,
                organizationId: input.organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
            };
            const stringifiedBody = JSON.stringify(body);
            const stamp = await activeStamper.stamp(stringifiedBody);
            return {
                body: stringifiedBody,
                stamp: stamp,
                url: fullUrl,
            };
        };
        this.getPrivateKey = async (input, stampWith) => {
            const session = await this.storageManager?.getActiveSession();
            return this.request("/public/v1/query/get_private_key", {
                ...input,
                organizationId: input.organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
            }, stampWith);
        };
        this.stampGetPrivateKey = async (input, stampWith) => {
            const activeStamper = this.getStamper(stampWith);
            if (!activeStamper) {
                return undefined;
            }
            const session = await this.storageManager?.getActiveSession();
            const fullUrl = this.config.apiBaseUrl + "/public/v1/query/get_private_key";
            const body = {
                ...input,
                organizationId: input.organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
            };
            const stringifiedBody = JSON.stringify(body);
            const stamp = await activeStamper.stamp(stringifiedBody);
            return {
                body: stringifiedBody,
                stamp: stamp,
                url: fullUrl,
            };
        };
        this.getSendTransactionStatus = async (input, stampWith) => {
            const session = await this.storageManager?.getActiveSession();
            return this.request("/public/v1/query/get_send_transaction_status", {
                ...input,
                organizationId: input.organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
            }, stampWith);
        };
        this.stampGetSendTransactionStatus = async (input, stampWith) => {
            const activeStamper = this.getStamper(stampWith);
            if (!activeStamper) {
                return undefined;
            }
            const session = await this.storageManager?.getActiveSession();
            const fullUrl = this.config.apiBaseUrl + "/public/v1/query/get_send_transaction_status";
            const body = {
                ...input,
                organizationId: input.organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
            };
            const stringifiedBody = JSON.stringify(body);
            const stamp = await activeStamper.stamp(stringifiedBody);
            return {
                body: stringifiedBody,
                stamp: stamp,
                url: fullUrl,
            };
        };
        this.getSmartContractInterface = async (input, stampWith) => {
            const session = await this.storageManager?.getActiveSession();
            return this.request("/public/v1/query/get_smart_contract_interface", {
                ...input,
                organizationId: input.organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
            }, stampWith);
        };
        this.stampGetSmartContractInterface = async (input, stampWith) => {
            const activeStamper = this.getStamper(stampWith);
            if (!activeStamper) {
                return undefined;
            }
            const session = await this.storageManager?.getActiveSession();
            const fullUrl = this.config.apiBaseUrl + "/public/v1/query/get_smart_contract_interface";
            const body = {
                ...input,
                organizationId: input.organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
            };
            const stringifiedBody = JSON.stringify(body);
            const stamp = await activeStamper.stamp(stringifiedBody);
            return {
                body: stringifiedBody,
                stamp: stamp,
                url: fullUrl,
            };
        };
        this.getUser = async (input, stampWith) => {
            const session = await this.storageManager?.getActiveSession();
            return this.request("/public/v1/query/get_user", {
                ...input,
                organizationId: input.organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
            }, stampWith);
        };
        this.stampGetUser = async (input, stampWith) => {
            const activeStamper = this.getStamper(stampWith);
            if (!activeStamper) {
                return undefined;
            }
            const session = await this.storageManager?.getActiveSession();
            const fullUrl = this.config.apiBaseUrl + "/public/v1/query/get_user";
            const body = {
                ...input,
                organizationId: input.organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
            };
            const stringifiedBody = JSON.stringify(body);
            const stamp = await activeStamper.stamp(stringifiedBody);
            return {
                body: stringifiedBody,
                stamp: stamp,
                url: fullUrl,
            };
        };
        this.getWallet = async (input, stampWith) => {
            const session = await this.storageManager?.getActiveSession();
            return this.request("/public/v1/query/get_wallet", {
                ...input,
                organizationId: input.organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
            }, stampWith);
        };
        this.stampGetWallet = async (input, stampWith) => {
            const activeStamper = this.getStamper(stampWith);
            if (!activeStamper) {
                return undefined;
            }
            const session = await this.storageManager?.getActiveSession();
            const fullUrl = this.config.apiBaseUrl + "/public/v1/query/get_wallet";
            const body = {
                ...input,
                organizationId: input.organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
            };
            const stringifiedBody = JSON.stringify(body);
            const stamp = await activeStamper.stamp(stringifiedBody);
            return {
                body: stringifiedBody,
                stamp: stamp,
                url: fullUrl,
            };
        };
        this.getWalletAccount = async (input, stampWith) => {
            const session = await this.storageManager?.getActiveSession();
            return this.request("/public/v1/query/get_wallet_account", {
                ...input,
                organizationId: input.organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
            }, stampWith);
        };
        this.stampGetWalletAccount = async (input, stampWith) => {
            const activeStamper = this.getStamper(stampWith);
            if (!activeStamper) {
                return undefined;
            }
            const session = await this.storageManager?.getActiveSession();
            const fullUrl = this.config.apiBaseUrl + "/public/v1/query/get_wallet_account";
            const body = {
                ...input,
                organizationId: input.organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
            };
            const stringifiedBody = JSON.stringify(body);
            const stamp = await activeStamper.stamp(stringifiedBody);
            return {
                body: stringifiedBody,
                stamp: stamp,
                url: fullUrl,
            };
        };
        this.getWalletAddressBalances = async (input, stampWith) => {
            const session = await this.storageManager?.getActiveSession();
            return this.request("/public/v1/query/get_wallet_address_balances", {
                ...input,
                organizationId: input.organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
            }, stampWith);
        };
        this.stampGetWalletAddressBalances = async (input, stampWith) => {
            const activeStamper = this.getStamper(stampWith);
            if (!activeStamper) {
                return undefined;
            }
            const session = await this.storageManager?.getActiveSession();
            const fullUrl = this.config.apiBaseUrl + "/public/v1/query/get_wallet_address_balances";
            const body = {
                ...input,
                organizationId: input.organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
            };
            const stringifiedBody = JSON.stringify(body);
            const stamp = await activeStamper.stamp(stringifiedBody);
            return {
                body: stringifiedBody,
                stamp: stamp,
                url: fullUrl,
            };
        };
        this.getActivities = async (input = {}, stampWith) => {
            const session = await this.storageManager?.getActiveSession();
            return this.request("/public/v1/query/list_activities", {
                ...input,
                organizationId: input.organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
            }, stampWith);
        };
        this.stampGetActivities = async (input, stampWith) => {
            const activeStamper = this.getStamper(stampWith);
            if (!activeStamper) {
                return undefined;
            }
            const session = await this.storageManager?.getActiveSession();
            const fullUrl = this.config.apiBaseUrl + "/public/v1/query/list_activities";
            const body = {
                ...input,
                organizationId: input.organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
            };
            const stringifiedBody = JSON.stringify(body);
            const stamp = await activeStamper.stamp(stringifiedBody);
            return {
                body: stringifiedBody,
                stamp: stamp,
                url: fullUrl,
            };
        };
        this.getAppProofs = async (input, stampWith) => {
            const session = await this.storageManager?.getActiveSession();
            return this.request("/public/v1/query/list_app_proofs", {
                ...input,
                organizationId: input.organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
            }, stampWith);
        };
        this.stampGetAppProofs = async (input, stampWith) => {
            const activeStamper = this.getStamper(stampWith);
            if (!activeStamper) {
                return undefined;
            }
            const session = await this.storageManager?.getActiveSession();
            const fullUrl = this.config.apiBaseUrl + "/public/v1/query/list_app_proofs";
            const body = {
                ...input,
                organizationId: input.organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
            };
            const stringifiedBody = JSON.stringify(body);
            const stamp = await activeStamper.stamp(stringifiedBody);
            return {
                body: stringifiedBody,
                stamp: stamp,
                url: fullUrl,
            };
        };
        this.listFiatOnRampCredentials = async (input, stampWith) => {
            const session = await this.storageManager?.getActiveSession();
            return this.request("/public/v1/query/list_fiat_on_ramp_credentials", {
                ...input,
                organizationId: input.organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
            }, stampWith);
        };
        this.stampListFiatOnRampCredentials = async (input, stampWith) => {
            const activeStamper = this.getStamper(stampWith);
            if (!activeStamper) {
                return undefined;
            }
            const session = await this.storageManager?.getActiveSession();
            const fullUrl = this.config.apiBaseUrl + "/public/v1/query/list_fiat_on_ramp_credentials";
            const body = {
                ...input,
                organizationId: input.organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
            };
            const stringifiedBody = JSON.stringify(body);
            const stamp = await activeStamper.stamp(stringifiedBody);
            return {
                body: stringifiedBody,
                stamp: stamp,
                url: fullUrl,
            };
        };
        this.listOauth2Credentials = async (input, stampWith) => {
            const session = await this.storageManager?.getActiveSession();
            return this.request("/public/v1/query/list_oauth2_credentials", {
                ...input,
                organizationId: input.organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
            }, stampWith);
        };
        this.stampListOauth2Credentials = async (input, stampWith) => {
            const activeStamper = this.getStamper(stampWith);
            if (!activeStamper) {
                return undefined;
            }
            const session = await this.storageManager?.getActiveSession();
            const fullUrl = this.config.apiBaseUrl + "/public/v1/query/list_oauth2_credentials";
            const body = {
                ...input,
                organizationId: input.organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
            };
            const stringifiedBody = JSON.stringify(body);
            const stamp = await activeStamper.stamp(stringifiedBody);
            return {
                body: stringifiedBody,
                stamp: stamp,
                url: fullUrl,
            };
        };
        this.getPolicies = async (input = {}, stampWith) => {
            const session = await this.storageManager?.getActiveSession();
            return this.request("/public/v1/query/list_policies", {
                ...input,
                organizationId: input.organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
            }, stampWith);
        };
        this.stampGetPolicies = async (input, stampWith) => {
            const activeStamper = this.getStamper(stampWith);
            if (!activeStamper) {
                return undefined;
            }
            const session = await this.storageManager?.getActiveSession();
            const fullUrl = this.config.apiBaseUrl + "/public/v1/query/list_policies";
            const body = {
                ...input,
                organizationId: input.organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
            };
            const stringifiedBody = JSON.stringify(body);
            const stamp = await activeStamper.stamp(stringifiedBody);
            return {
                body: stringifiedBody,
                stamp: stamp,
                url: fullUrl,
            };
        };
        this.listPrivateKeyTags = async (input, stampWith) => {
            const session = await this.storageManager?.getActiveSession();
            return this.request("/public/v1/query/list_private_key_tags", {
                ...input,
                organizationId: input.organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
            }, stampWith);
        };
        this.stampListPrivateKeyTags = async (input, stampWith) => {
            const activeStamper = this.getStamper(stampWith);
            if (!activeStamper) {
                return undefined;
            }
            const session = await this.storageManager?.getActiveSession();
            const fullUrl = this.config.apiBaseUrl + "/public/v1/query/list_private_key_tags";
            const body = {
                ...input,
                organizationId: input.organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
            };
            const stringifiedBody = JSON.stringify(body);
            const stamp = await activeStamper.stamp(stringifiedBody);
            return {
                body: stringifiedBody,
                stamp: stamp,
                url: fullUrl,
            };
        };
        this.getPrivateKeys = async (input = {}, stampWith) => {
            const session = await this.storageManager?.getActiveSession();
            return this.request("/public/v1/query/list_private_keys", {
                ...input,
                organizationId: input.organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
            }, stampWith);
        };
        this.stampGetPrivateKeys = async (input, stampWith) => {
            const activeStamper = this.getStamper(stampWith);
            if (!activeStamper) {
                return undefined;
            }
            const session = await this.storageManager?.getActiveSession();
            const fullUrl = this.config.apiBaseUrl + "/public/v1/query/list_private_keys";
            const body = {
                ...input,
                organizationId: input.organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
            };
            const stringifiedBody = JSON.stringify(body);
            const stamp = await activeStamper.stamp(stringifiedBody);
            return {
                body: stringifiedBody,
                stamp: stamp,
                url: fullUrl,
            };
        };
        this.getSmartContractInterfaces = async (input, stampWith) => {
            const session = await this.storageManager?.getActiveSession();
            return this.request("/public/v1/query/list_smart_contract_interfaces", {
                ...input,
                organizationId: input.organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
            }, stampWith);
        };
        this.stampGetSmartContractInterfaces = async (input, stampWith) => {
            const activeStamper = this.getStamper(stampWith);
            if (!activeStamper) {
                return undefined;
            }
            const session = await this.storageManager?.getActiveSession();
            const fullUrl = this.config.apiBaseUrl +
                "/public/v1/query/list_smart_contract_interfaces";
            const body = {
                ...input,
                organizationId: input.organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
            };
            const stringifiedBody = JSON.stringify(body);
            const stamp = await activeStamper.stamp(stringifiedBody);
            return {
                body: stringifiedBody,
                stamp: stamp,
                url: fullUrl,
            };
        };
        this.getSubOrgIds = async (input = {}, stampWith) => {
            const session = await this.storageManager?.getActiveSession();
            return this.request("/public/v1/query/list_suborgs", {
                ...input,
                organizationId: input.organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
            }, stampWith);
        };
        this.stampGetSubOrgIds = async (input, stampWith) => {
            const activeStamper = this.getStamper(stampWith);
            if (!activeStamper) {
                return undefined;
            }
            const session = await this.storageManager?.getActiveSession();
            const fullUrl = this.config.apiBaseUrl + "/public/v1/query/list_suborgs";
            const body = {
                ...input,
                organizationId: input.organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
            };
            const stringifiedBody = JSON.stringify(body);
            const stamp = await activeStamper.stamp(stringifiedBody);
            return {
                body: stringifiedBody,
                stamp: stamp,
                url: fullUrl,
            };
        };
        this.listSupportedAssets = async (input, stampWith) => {
            const session = await this.storageManager?.getActiveSession();
            return this.request("/public/v1/query/list_supported_assets", {
                ...input,
                organizationId: input.organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
            }, stampWith);
        };
        this.stampListSupportedAssets = async (input, stampWith) => {
            const activeStamper = this.getStamper(stampWith);
            if (!activeStamper) {
                return undefined;
            }
            const session = await this.storageManager?.getActiveSession();
            const fullUrl = this.config.apiBaseUrl + "/public/v1/query/list_supported_assets";
            const body = {
                ...input,
                organizationId: input.organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
            };
            const stringifiedBody = JSON.stringify(body);
            const stamp = await activeStamper.stamp(stringifiedBody);
            return {
                body: stringifiedBody,
                stamp: stamp,
                url: fullUrl,
            };
        };
        this.listUserTags = async (input = {}, stampWith) => {
            const session = await this.storageManager?.getActiveSession();
            return this.request("/public/v1/query/list_user_tags", {
                ...input,
                organizationId: input.organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
            }, stampWith);
        };
        this.stampListUserTags = async (input, stampWith) => {
            const activeStamper = this.getStamper(stampWith);
            if (!activeStamper) {
                return undefined;
            }
            const session = await this.storageManager?.getActiveSession();
            const fullUrl = this.config.apiBaseUrl + "/public/v1/query/list_user_tags";
            const body = {
                ...input,
                organizationId: input.organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
            };
            const stringifiedBody = JSON.stringify(body);
            const stamp = await activeStamper.stamp(stringifiedBody);
            return {
                body: stringifiedBody,
                stamp: stamp,
                url: fullUrl,
            };
        };
        this.getUsers = async (input = {}, stampWith) => {
            const session = await this.storageManager?.getActiveSession();
            return this.request("/public/v1/query/list_users", {
                ...input,
                organizationId: input.organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
            }, stampWith);
        };
        this.stampGetUsers = async (input, stampWith) => {
            const activeStamper = this.getStamper(stampWith);
            if (!activeStamper) {
                return undefined;
            }
            const session = await this.storageManager?.getActiveSession();
            const fullUrl = this.config.apiBaseUrl + "/public/v1/query/list_users";
            const body = {
                ...input,
                organizationId: input.organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
            };
            const stringifiedBody = JSON.stringify(body);
            const stamp = await activeStamper.stamp(stringifiedBody);
            return {
                body: stringifiedBody,
                stamp: stamp,
                url: fullUrl,
            };
        };
        this.getVerifiedSubOrgIds = async (input, stampWith) => {
            const session = await this.storageManager?.getActiveSession();
            return this.request("/public/v1/query/list_verified_suborgs", {
                ...input,
                organizationId: input.organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
            }, stampWith);
        };
        this.stampGetVerifiedSubOrgIds = async (input, stampWith) => {
            const activeStamper = this.getStamper(stampWith);
            if (!activeStamper) {
                return undefined;
            }
            const session = await this.storageManager?.getActiveSession();
            const fullUrl = this.config.apiBaseUrl + "/public/v1/query/list_verified_suborgs";
            const body = {
                ...input,
                organizationId: input.organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
            };
            const stringifiedBody = JSON.stringify(body);
            const stamp = await activeStamper.stamp(stringifiedBody);
            return {
                body: stringifiedBody,
                stamp: stamp,
                url: fullUrl,
            };
        };
        this.getWalletAccounts = async (input, stampWith) => {
            const session = await this.storageManager?.getActiveSession();
            return this.request("/public/v1/query/list_wallet_accounts", {
                ...input,
                organizationId: input.organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
            }, stampWith);
        };
        this.stampGetWalletAccounts = async (input, stampWith) => {
            const activeStamper = this.getStamper(stampWith);
            if (!activeStamper) {
                return undefined;
            }
            const session = await this.storageManager?.getActiveSession();
            const fullUrl = this.config.apiBaseUrl + "/public/v1/query/list_wallet_accounts";
            const body = {
                ...input,
                organizationId: input.organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
            };
            const stringifiedBody = JSON.stringify(body);
            const stamp = await activeStamper.stamp(stringifiedBody);
            return {
                body: stringifiedBody,
                stamp: stamp,
                url: fullUrl,
            };
        };
        this.getWallets = async (input = {}, stampWith) => {
            const session = await this.storageManager?.getActiveSession();
            return this.request("/public/v1/query/list_wallets", {
                ...input,
                organizationId: input.organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
            }, stampWith);
        };
        this.stampGetWallets = async (input, stampWith) => {
            const activeStamper = this.getStamper(stampWith);
            if (!activeStamper) {
                return undefined;
            }
            const session = await this.storageManager?.getActiveSession();
            const fullUrl = this.config.apiBaseUrl + "/public/v1/query/list_wallets";
            const body = {
                ...input,
                organizationId: input.organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
            };
            const stringifiedBody = JSON.stringify(body);
            const stamp = await activeStamper.stamp(stringifiedBody);
            return {
                body: stringifiedBody,
                stamp: stamp,
                url: fullUrl,
            };
        };
        this.getWhoami = async (input = {}, stampWith) => {
            const session = await this.storageManager?.getActiveSession();
            return this.request("/public/v1/query/whoami", {
                ...input,
                organizationId: input.organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
            }, stampWith);
        };
        this.stampGetWhoami = async (input, stampWith) => {
            const activeStamper = this.getStamper(stampWith);
            if (!activeStamper) {
                return undefined;
            }
            const session = await this.storageManager?.getActiveSession();
            const fullUrl = this.config.apiBaseUrl + "/public/v1/query/whoami";
            const body = {
                ...input,
                organizationId: input.organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
            };
            const stringifiedBody = JSON.stringify(body);
            const stamp = await activeStamper.stamp(stringifiedBody);
            return {
                body: stringifiedBody,
                stamp: stamp,
                url: fullUrl,
            };
        };
        this.approveActivity = async (input, stampWith) => {
            const { organizationId, timestampMs, ...rest } = input;
            const session = await this.storageManager?.getActiveSession();
            return this.activityDecision("/public/v1/submit/approve_activity", {
                parameters: rest,
                organizationId: organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_APPROVE_ACTIVITY",
            }, stampWith);
        };
        this.stampApproveActivity = async (input, stampWith) => {
            const activeStamper = this.getStamper(stampWith);
            if (!activeStamper) {
                return undefined;
            }
            const { organizationId, timestampMs, ...parameters } = input;
            const session = await this.storageManager?.getActiveSession();
            const fullUrl = this.config.apiBaseUrl + "/public/v1/submit/approve_activity";
            const bodyWithType = {
                parameters,
                organizationId: organizationId ?? session?.organizationId ?? this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_APPROVE_ACTIVITY",
            };
            const stringifiedBody = JSON.stringify(bodyWithType);
            const stamp = await activeStamper.stamp(stringifiedBody);
            return {
                body: stringifiedBody,
                stamp: stamp,
                url: fullUrl,
            };
        };
        this.createApiKeys = async (input, stampWith) => {
            const { organizationId, timestampMs, ...rest } = input;
            const session = await this.storageManager?.getActiveSession();
            return this.activity("/public/v1/submit/create_api_keys", {
                parameters: rest,
                organizationId: organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_CREATE_API_KEYS_V2",
            }, "createApiKeysResult", stampWith);
        };
        this.stampCreateApiKeys = async (input, stampWith) => {
            const activeStamper = this.getStamper(stampWith);
            if (!activeStamper) {
                return undefined;
            }
            const { organizationId, timestampMs, ...parameters } = input;
            const session = await this.storageManager?.getActiveSession();
            const fullUrl = this.config.apiBaseUrl + "/public/v1/submit/create_api_keys";
            const bodyWithType = {
                parameters,
                organizationId: organizationId ?? session?.organizationId ?? this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_CREATE_API_KEYS_V2",
            };
            const stringifiedBody = JSON.stringify(bodyWithType);
            const stamp = await activeStamper.stamp(stringifiedBody);
            return {
                body: stringifiedBody,
                stamp: stamp,
                url: fullUrl,
            };
        };
        this.createAuthenticators = async (input, stampWith) => {
            const { organizationId, timestampMs, ...rest } = input;
            const session = await this.storageManager?.getActiveSession();
            return this.activity("/public/v1/submit/create_authenticators", {
                parameters: rest,
                organizationId: organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_CREATE_AUTHENTICATORS_V2",
            }, "createAuthenticatorsResult", stampWith);
        };
        this.stampCreateAuthenticators = async (input, stampWith) => {
            const activeStamper = this.getStamper(stampWith);
            if (!activeStamper) {
                return undefined;
            }
            const { organizationId, timestampMs, ...parameters } = input;
            const session = await this.storageManager?.getActiveSession();
            const fullUrl = this.config.apiBaseUrl + "/public/v1/submit/create_authenticators";
            const bodyWithType = {
                parameters,
                organizationId: organizationId ?? session?.organizationId ?? this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_CREATE_AUTHENTICATORS_V2",
            };
            const stringifiedBody = JSON.stringify(bodyWithType);
            const stamp = await activeStamper.stamp(stringifiedBody);
            return {
                body: stringifiedBody,
                stamp: stamp,
                url: fullUrl,
            };
        };
        this.createFiatOnRampCredential = async (input, stampWith) => {
            const { organizationId, timestampMs, ...rest } = input;
            const session = await this.storageManager?.getActiveSession();
            return this.activity("/public/v1/submit/create_fiat_on_ramp_credential", {
                parameters: rest,
                organizationId: organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_CREATE_FIAT_ON_RAMP_CREDENTIAL",
            }, "createFiatOnRampCredentialResult", stampWith);
        };
        this.stampCreateFiatOnRampCredential = async (input, stampWith) => {
            const activeStamper = this.getStamper(stampWith);
            if (!activeStamper) {
                return undefined;
            }
            const { organizationId, timestampMs, ...parameters } = input;
            const session = await this.storageManager?.getActiveSession();
            const fullUrl = this.config.apiBaseUrl +
                "/public/v1/submit/create_fiat_on_ramp_credential";
            const bodyWithType = {
                parameters,
                organizationId: organizationId ?? session?.organizationId ?? this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_CREATE_FIAT_ON_RAMP_CREDENTIAL",
            };
            const stringifiedBody = JSON.stringify(bodyWithType);
            const stamp = await activeStamper.stamp(stringifiedBody);
            return {
                body: stringifiedBody,
                stamp: stamp,
                url: fullUrl,
            };
        };
        this.createInvitations = async (input, stampWith) => {
            const { organizationId, timestampMs, ...rest } = input;
            const session = await this.storageManager?.getActiveSession();
            return this.activity("/public/v1/submit/create_invitations", {
                parameters: rest,
                organizationId: organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_CREATE_INVITATIONS",
            }, "createInvitationsResult", stampWith);
        };
        this.stampCreateInvitations = async (input, stampWith) => {
            const activeStamper = this.getStamper(stampWith);
            if (!activeStamper) {
                return undefined;
            }
            const { organizationId, timestampMs, ...parameters } = input;
            const session = await this.storageManager?.getActiveSession();
            const fullUrl = this.config.apiBaseUrl + "/public/v1/submit/create_invitations";
            const bodyWithType = {
                parameters,
                organizationId: organizationId ?? session?.organizationId ?? this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_CREATE_INVITATIONS",
            };
            const stringifiedBody = JSON.stringify(bodyWithType);
            const stamp = await activeStamper.stamp(stringifiedBody);
            return {
                body: stringifiedBody,
                stamp: stamp,
                url: fullUrl,
            };
        };
        this.createOauth2Credential = async (input, stampWith) => {
            const { organizationId, timestampMs, ...rest } = input;
            const session = await this.storageManager?.getActiveSession();
            return this.activity("/public/v1/submit/create_oauth2_credential", {
                parameters: rest,
                organizationId: organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_CREATE_OAUTH2CREDENTIAL",
            }, "createOauth2CredentialResult", stampWith);
        };
        this.stampCreateOauth2Credential = async (input, stampWith) => {
            const activeStamper = this.getStamper(stampWith);
            if (!activeStamper) {
                return undefined;
            }
            const { organizationId, timestampMs, ...parameters } = input;
            const session = await this.storageManager?.getActiveSession();
            const fullUrl = this.config.apiBaseUrl + "/public/v1/submit/create_oauth2_credential";
            const bodyWithType = {
                parameters,
                organizationId: organizationId ?? session?.organizationId ?? this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_CREATE_OAUTH2CREDENTIAL",
            };
            const stringifiedBody = JSON.stringify(bodyWithType);
            const stamp = await activeStamper.stamp(stringifiedBody);
            return {
                body: stringifiedBody,
                stamp: stamp,
                url: fullUrl,
            };
        };
        this.createOauthProviders = async (input, stampWith) => {
            const { organizationId, timestampMs, ...rest } = input;
            const session = await this.storageManager?.getActiveSession();
            return this.activity("/public/v1/submit/create_oauth_providers", {
                parameters: rest,
                organizationId: organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_CREATE_OAUTH_PROVIDERS",
            }, "createOauthProvidersResult", stampWith);
        };
        this.stampCreateOauthProviders = async (input, stampWith) => {
            const activeStamper = this.getStamper(stampWith);
            if (!activeStamper) {
                return undefined;
            }
            const { organizationId, timestampMs, ...parameters } = input;
            const session = await this.storageManager?.getActiveSession();
            const fullUrl = this.config.apiBaseUrl + "/public/v1/submit/create_oauth_providers";
            const bodyWithType = {
                parameters,
                organizationId: organizationId ?? session?.organizationId ?? this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_CREATE_OAUTH_PROVIDERS",
            };
            const stringifiedBody = JSON.stringify(bodyWithType);
            const stamp = await activeStamper.stamp(stringifiedBody);
            return {
                body: stringifiedBody,
                stamp: stamp,
                url: fullUrl,
            };
        };
        this.createPolicies = async (input, stampWith) => {
            const { organizationId, timestampMs, ...rest } = input;
            const session = await this.storageManager?.getActiveSession();
            return this.activity("/public/v1/submit/create_policies", {
                parameters: rest,
                organizationId: organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_CREATE_POLICIES",
            }, "createPoliciesResult", stampWith);
        };
        this.stampCreatePolicies = async (input, stampWith) => {
            const activeStamper = this.getStamper(stampWith);
            if (!activeStamper) {
                return undefined;
            }
            const { organizationId, timestampMs, ...parameters } = input;
            const session = await this.storageManager?.getActiveSession();
            const fullUrl = this.config.apiBaseUrl + "/public/v1/submit/create_policies";
            const bodyWithType = {
                parameters,
                organizationId: organizationId ?? session?.organizationId ?? this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_CREATE_POLICIES",
            };
            const stringifiedBody = JSON.stringify(bodyWithType);
            const stamp = await activeStamper.stamp(stringifiedBody);
            return {
                body: stringifiedBody,
                stamp: stamp,
                url: fullUrl,
            };
        };
        this.createPolicy = async (input, stampWith) => {
            const { organizationId, timestampMs, ...rest } = input;
            const session = await this.storageManager?.getActiveSession();
            return this.activity("/public/v1/submit/create_policy", {
                parameters: rest,
                organizationId: organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_CREATE_POLICY_V3",
            }, "createPolicyResult", stampWith);
        };
        this.stampCreatePolicy = async (input, stampWith) => {
            const activeStamper = this.getStamper(stampWith);
            if (!activeStamper) {
                return undefined;
            }
            const { organizationId, timestampMs, ...parameters } = input;
            const session = await this.storageManager?.getActiveSession();
            const fullUrl = this.config.apiBaseUrl + "/public/v1/submit/create_policy";
            const bodyWithType = {
                parameters,
                organizationId: organizationId ?? session?.organizationId ?? this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_CREATE_POLICY_V3",
            };
            const stringifiedBody = JSON.stringify(bodyWithType);
            const stamp = await activeStamper.stamp(stringifiedBody);
            return {
                body: stringifiedBody,
                stamp: stamp,
                url: fullUrl,
            };
        };
        this.createPrivateKeyTag = async (input, stampWith) => {
            const { organizationId, timestampMs, ...rest } = input;
            const session = await this.storageManager?.getActiveSession();
            return this.activity("/public/v1/submit/create_private_key_tag", {
                parameters: rest,
                organizationId: organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_CREATE_PRIVATE_KEY_TAG",
            }, "createPrivateKeyTagResult", stampWith);
        };
        this.stampCreatePrivateKeyTag = async (input, stampWith) => {
            const activeStamper = this.getStamper(stampWith);
            if (!activeStamper) {
                return undefined;
            }
            const { organizationId, timestampMs, ...parameters } = input;
            const session = await this.storageManager?.getActiveSession();
            const fullUrl = this.config.apiBaseUrl + "/public/v1/submit/create_private_key_tag";
            const bodyWithType = {
                parameters,
                organizationId: organizationId ?? session?.organizationId ?? this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_CREATE_PRIVATE_KEY_TAG",
            };
            const stringifiedBody = JSON.stringify(bodyWithType);
            const stamp = await activeStamper.stamp(stringifiedBody);
            return {
                body: stringifiedBody,
                stamp: stamp,
                url: fullUrl,
            };
        };
        this.createPrivateKeys = async (input, stampWith) => {
            const { organizationId, timestampMs, ...rest } = input;
            const session = await this.storageManager?.getActiveSession();
            return this.activity("/public/v1/submit/create_private_keys", {
                parameters: rest,
                organizationId: organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_CREATE_PRIVATE_KEYS_V2",
            }, "createPrivateKeysResultV2", stampWith);
        };
        this.stampCreatePrivateKeys = async (input, stampWith) => {
            const activeStamper = this.getStamper(stampWith);
            if (!activeStamper) {
                return undefined;
            }
            const { organizationId, timestampMs, ...parameters } = input;
            const session = await this.storageManager?.getActiveSession();
            const fullUrl = this.config.apiBaseUrl + "/public/v1/submit/create_private_keys";
            const bodyWithType = {
                parameters,
                organizationId: organizationId ?? session?.organizationId ?? this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_CREATE_PRIVATE_KEYS_V2",
            };
            const stringifiedBody = JSON.stringify(bodyWithType);
            const stamp = await activeStamper.stamp(stringifiedBody);
            return {
                body: stringifiedBody,
                stamp: stamp,
                url: fullUrl,
            };
        };
        this.createReadOnlySession = async (input, stampWith) => {
            const { organizationId, timestampMs, ...rest } = input;
            const session = await this.storageManager?.getActiveSession();
            return this.activity("/public/v1/submit/create_read_only_session", {
                parameters: rest,
                organizationId: organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_CREATE_READ_ONLY_SESSION",
            }, "createReadOnlySessionResult", stampWith);
        };
        this.stampCreateReadOnlySession = async (input, stampWith) => {
            const activeStamper = this.getStamper(stampWith);
            if (!activeStamper) {
                return undefined;
            }
            const { organizationId, timestampMs, ...parameters } = input;
            const session = await this.storageManager?.getActiveSession();
            const fullUrl = this.config.apiBaseUrl + "/public/v1/submit/create_read_only_session";
            const bodyWithType = {
                parameters,
                organizationId: organizationId ?? session?.organizationId ?? this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_CREATE_READ_ONLY_SESSION",
            };
            const stringifiedBody = JSON.stringify(bodyWithType);
            const stamp = await activeStamper.stamp(stringifiedBody);
            return {
                body: stringifiedBody,
                stamp: stamp,
                url: fullUrl,
            };
        };
        this.createReadWriteSession = async (input, stampWith) => {
            const { organizationId, timestampMs, ...rest } = input;
            const session = await this.storageManager?.getActiveSession();
            return this.activity("/public/v1/submit/create_read_write_session", {
                parameters: rest,
                organizationId: organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_CREATE_READ_WRITE_SESSION_V2",
            }, "createReadWriteSessionResultV2", stampWith);
        };
        this.stampCreateReadWriteSession = async (input, stampWith) => {
            const activeStamper = this.getStamper(stampWith);
            if (!activeStamper) {
                return undefined;
            }
            const { organizationId, timestampMs, ...parameters } = input;
            const session = await this.storageManager?.getActiveSession();
            const fullUrl = this.config.apiBaseUrl + "/public/v1/submit/create_read_write_session";
            const bodyWithType = {
                parameters,
                organizationId: organizationId ?? session?.organizationId ?? this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_CREATE_READ_WRITE_SESSION_V2",
            };
            const stringifiedBody = JSON.stringify(bodyWithType);
            const stamp = await activeStamper.stamp(stringifiedBody);
            return {
                body: stringifiedBody,
                stamp: stamp,
                url: fullUrl,
            };
        };
        this.createSmartContractInterface = async (input, stampWith) => {
            const { organizationId, timestampMs, ...rest } = input;
            const session = await this.storageManager?.getActiveSession();
            return this.activity("/public/v1/submit/create_smart_contract_interface", {
                parameters: rest,
                organizationId: organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_CREATE_SMART_CONTRACT_INTERFACE",
            }, "createSmartContractInterfaceResult", stampWith);
        };
        this.stampCreateSmartContractInterface = async (input, stampWith) => {
            const activeStamper = this.getStamper(stampWith);
            if (!activeStamper) {
                return undefined;
            }
            const { organizationId, timestampMs, ...parameters } = input;
            const session = await this.storageManager?.getActiveSession();
            const fullUrl = this.config.apiBaseUrl +
                "/public/v1/submit/create_smart_contract_interface";
            const bodyWithType = {
                parameters,
                organizationId: organizationId ?? session?.organizationId ?? this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_CREATE_SMART_CONTRACT_INTERFACE",
            };
            const stringifiedBody = JSON.stringify(bodyWithType);
            const stamp = await activeStamper.stamp(stringifiedBody);
            return {
                body: stringifiedBody,
                stamp: stamp,
                url: fullUrl,
            };
        };
        this.createSubOrganization = async (input, stampWith) => {
            const { organizationId, timestampMs, ...rest } = input;
            const session = await this.storageManager?.getActiveSession();
            return this.activity("/public/v1/submit/create_sub_organization", {
                parameters: rest,
                organizationId: organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_CREATE_SUB_ORGANIZATION_V7",
            }, "createSubOrganizationResultV7", stampWith);
        };
        this.stampCreateSubOrganization = async (input, stampWith) => {
            const activeStamper = this.getStamper(stampWith);
            if (!activeStamper) {
                return undefined;
            }
            const { organizationId, timestampMs, ...parameters } = input;
            const session = await this.storageManager?.getActiveSession();
            const fullUrl = this.config.apiBaseUrl + "/public/v1/submit/create_sub_organization";
            const bodyWithType = {
                parameters,
                organizationId: organizationId ?? session?.organizationId ?? this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_CREATE_SUB_ORGANIZATION_V7",
            };
            const stringifiedBody = JSON.stringify(bodyWithType);
            const stamp = await activeStamper.stamp(stringifiedBody);
            return {
                body: stringifiedBody,
                stamp: stamp,
                url: fullUrl,
            };
        };
        this.createUserTag = async (input, stampWith) => {
            const { organizationId, timestampMs, ...rest } = input;
            const session = await this.storageManager?.getActiveSession();
            return this.activity("/public/v1/submit/create_user_tag", {
                parameters: rest,
                organizationId: organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_CREATE_USER_TAG",
            }, "createUserTagResult", stampWith);
        };
        this.stampCreateUserTag = async (input, stampWith) => {
            const activeStamper = this.getStamper(stampWith);
            if (!activeStamper) {
                return undefined;
            }
            const { organizationId, timestampMs, ...parameters } = input;
            const session = await this.storageManager?.getActiveSession();
            const fullUrl = this.config.apiBaseUrl + "/public/v1/submit/create_user_tag";
            const bodyWithType = {
                parameters,
                organizationId: organizationId ?? session?.organizationId ?? this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_CREATE_USER_TAG",
            };
            const stringifiedBody = JSON.stringify(bodyWithType);
            const stamp = await activeStamper.stamp(stringifiedBody);
            return {
                body: stringifiedBody,
                stamp: stamp,
                url: fullUrl,
            };
        };
        this.createUsers = async (input, stampWith) => {
            const { organizationId, timestampMs, ...rest } = input;
            const session = await this.storageManager?.getActiveSession();
            return this.activity("/public/v1/submit/create_users", {
                parameters: rest,
                organizationId: organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_CREATE_USERS_V3",
            }, "createUsersResult", stampWith);
        };
        this.stampCreateUsers = async (input, stampWith) => {
            const activeStamper = this.getStamper(stampWith);
            if (!activeStamper) {
                return undefined;
            }
            const { organizationId, timestampMs, ...parameters } = input;
            const session = await this.storageManager?.getActiveSession();
            const fullUrl = this.config.apiBaseUrl + "/public/v1/submit/create_users";
            const bodyWithType = {
                parameters,
                organizationId: organizationId ?? session?.organizationId ?? this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_CREATE_USERS_V3",
            };
            const stringifiedBody = JSON.stringify(bodyWithType);
            const stamp = await activeStamper.stamp(stringifiedBody);
            return {
                body: stringifiedBody,
                stamp: stamp,
                url: fullUrl,
            };
        };
        this.createWallet = async (input, stampWith) => {
            const { organizationId, timestampMs, ...rest } = input;
            const session = await this.storageManager?.getActiveSession();
            return this.activity("/public/v1/submit/create_wallet", {
                parameters: rest,
                organizationId: organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_CREATE_WALLET",
            }, "createWalletResult", stampWith);
        };
        this.stampCreateWallet = async (input, stampWith) => {
            const activeStamper = this.getStamper(stampWith);
            if (!activeStamper) {
                return undefined;
            }
            const { organizationId, timestampMs, ...parameters } = input;
            const session = await this.storageManager?.getActiveSession();
            const fullUrl = this.config.apiBaseUrl + "/public/v1/submit/create_wallet";
            const bodyWithType = {
                parameters,
                organizationId: organizationId ?? session?.organizationId ?? this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_CREATE_WALLET",
            };
            const stringifiedBody = JSON.stringify(bodyWithType);
            const stamp = await activeStamper.stamp(stringifiedBody);
            return {
                body: stringifiedBody,
                stamp: stamp,
                url: fullUrl,
            };
        };
        this.createWalletAccounts = async (input, stampWith) => {
            const { organizationId, timestampMs, ...rest } = input;
            const session = await this.storageManager?.getActiveSession();
            return this.activity("/public/v1/submit/create_wallet_accounts", {
                parameters: rest,
                organizationId: organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_CREATE_WALLET_ACCOUNTS",
            }, "createWalletAccountsResult", stampWith);
        };
        this.stampCreateWalletAccounts = async (input, stampWith) => {
            const activeStamper = this.getStamper(stampWith);
            if (!activeStamper) {
                return undefined;
            }
            const { organizationId, timestampMs, ...parameters } = input;
            const session = await this.storageManager?.getActiveSession();
            const fullUrl = this.config.apiBaseUrl + "/public/v1/submit/create_wallet_accounts";
            const bodyWithType = {
                parameters,
                organizationId: organizationId ?? session?.organizationId ?? this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_CREATE_WALLET_ACCOUNTS",
            };
            const stringifiedBody = JSON.stringify(bodyWithType);
            const stamp = await activeStamper.stamp(stringifiedBody);
            return {
                body: stringifiedBody,
                stamp: stamp,
                url: fullUrl,
            };
        };
        this.deleteApiKeys = async (input, stampWith) => {
            const { organizationId, timestampMs, ...rest } = input;
            const session = await this.storageManager?.getActiveSession();
            return this.activity("/public/v1/submit/delete_api_keys", {
                parameters: rest,
                organizationId: organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_DELETE_API_KEYS",
            }, "deleteApiKeysResult", stampWith);
        };
        this.stampDeleteApiKeys = async (input, stampWith) => {
            const activeStamper = this.getStamper(stampWith);
            if (!activeStamper) {
                return undefined;
            }
            const { organizationId, timestampMs, ...parameters } = input;
            const session = await this.storageManager?.getActiveSession();
            const fullUrl = this.config.apiBaseUrl + "/public/v1/submit/delete_api_keys";
            const bodyWithType = {
                parameters,
                organizationId: organizationId ?? session?.organizationId ?? this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_DELETE_API_KEYS",
            };
            const stringifiedBody = JSON.stringify(bodyWithType);
            const stamp = await activeStamper.stamp(stringifiedBody);
            return {
                body: stringifiedBody,
                stamp: stamp,
                url: fullUrl,
            };
        };
        this.deleteAuthenticators = async (input, stampWith) => {
            const { organizationId, timestampMs, ...rest } = input;
            const session = await this.storageManager?.getActiveSession();
            return this.activity("/public/v1/submit/delete_authenticators", {
                parameters: rest,
                organizationId: organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_DELETE_AUTHENTICATORS",
            }, "deleteAuthenticatorsResult", stampWith);
        };
        this.stampDeleteAuthenticators = async (input, stampWith) => {
            const activeStamper = this.getStamper(stampWith);
            if (!activeStamper) {
                return undefined;
            }
            const { organizationId, timestampMs, ...parameters } = input;
            const session = await this.storageManager?.getActiveSession();
            const fullUrl = this.config.apiBaseUrl + "/public/v1/submit/delete_authenticators";
            const bodyWithType = {
                parameters,
                organizationId: organizationId ?? session?.organizationId ?? this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_DELETE_AUTHENTICATORS",
            };
            const stringifiedBody = JSON.stringify(bodyWithType);
            const stamp = await activeStamper.stamp(stringifiedBody);
            return {
                body: stringifiedBody,
                stamp: stamp,
                url: fullUrl,
            };
        };
        this.deleteFiatOnRampCredential = async (input, stampWith) => {
            const { organizationId, timestampMs, ...rest } = input;
            const session = await this.storageManager?.getActiveSession();
            return this.activity("/public/v1/submit/delete_fiat_on_ramp_credential", {
                parameters: rest,
                organizationId: organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_DELETE_FIAT_ON_RAMP_CREDENTIAL",
            }, "deleteFiatOnRampCredentialResult", stampWith);
        };
        this.stampDeleteFiatOnRampCredential = async (input, stampWith) => {
            const activeStamper = this.getStamper(stampWith);
            if (!activeStamper) {
                return undefined;
            }
            const { organizationId, timestampMs, ...parameters } = input;
            const session = await this.storageManager?.getActiveSession();
            const fullUrl = this.config.apiBaseUrl +
                "/public/v1/submit/delete_fiat_on_ramp_credential";
            const bodyWithType = {
                parameters,
                organizationId: organizationId ?? session?.organizationId ?? this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_DELETE_FIAT_ON_RAMP_CREDENTIAL",
            };
            const stringifiedBody = JSON.stringify(bodyWithType);
            const stamp = await activeStamper.stamp(stringifiedBody);
            return {
                body: stringifiedBody,
                stamp: stamp,
                url: fullUrl,
            };
        };
        this.deleteInvitation = async (input, stampWith) => {
            const { organizationId, timestampMs, ...rest } = input;
            const session = await this.storageManager?.getActiveSession();
            return this.activity("/public/v1/submit/delete_invitation", {
                parameters: rest,
                organizationId: organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_DELETE_INVITATION",
            }, "deleteInvitationResult", stampWith);
        };
        this.stampDeleteInvitation = async (input, stampWith) => {
            const activeStamper = this.getStamper(stampWith);
            if (!activeStamper) {
                return undefined;
            }
            const { organizationId, timestampMs, ...parameters } = input;
            const session = await this.storageManager?.getActiveSession();
            const fullUrl = this.config.apiBaseUrl + "/public/v1/submit/delete_invitation";
            const bodyWithType = {
                parameters,
                organizationId: organizationId ?? session?.organizationId ?? this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_DELETE_INVITATION",
            };
            const stringifiedBody = JSON.stringify(bodyWithType);
            const stamp = await activeStamper.stamp(stringifiedBody);
            return {
                body: stringifiedBody,
                stamp: stamp,
                url: fullUrl,
            };
        };
        this.deleteOauth2Credential = async (input, stampWith) => {
            const { organizationId, timestampMs, ...rest } = input;
            const session = await this.storageManager?.getActiveSession();
            return this.activity("/public/v1/submit/delete_oauth2_credential", {
                parameters: rest,
                organizationId: organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_DELETE_OAUTH2CREDENTIAL",
            }, "deleteOauth2CredentialResult", stampWith);
        };
        this.stampDeleteOauth2Credential = async (input, stampWith) => {
            const activeStamper = this.getStamper(stampWith);
            if (!activeStamper) {
                return undefined;
            }
            const { organizationId, timestampMs, ...parameters } = input;
            const session = await this.storageManager?.getActiveSession();
            const fullUrl = this.config.apiBaseUrl + "/public/v1/submit/delete_oauth2_credential";
            const bodyWithType = {
                parameters,
                organizationId: organizationId ?? session?.organizationId ?? this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_DELETE_OAUTH2CREDENTIAL",
            };
            const stringifiedBody = JSON.stringify(bodyWithType);
            const stamp = await activeStamper.stamp(stringifiedBody);
            return {
                body: stringifiedBody,
                stamp: stamp,
                url: fullUrl,
            };
        };
        this.deleteOauthProviders = async (input, stampWith) => {
            const { organizationId, timestampMs, ...rest } = input;
            const session = await this.storageManager?.getActiveSession();
            return this.activity("/public/v1/submit/delete_oauth_providers", {
                parameters: rest,
                organizationId: organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_DELETE_OAUTH_PROVIDERS",
            }, "deleteOauthProvidersResult", stampWith);
        };
        this.stampDeleteOauthProviders = async (input, stampWith) => {
            const activeStamper = this.getStamper(stampWith);
            if (!activeStamper) {
                return undefined;
            }
            const { organizationId, timestampMs, ...parameters } = input;
            const session = await this.storageManager?.getActiveSession();
            const fullUrl = this.config.apiBaseUrl + "/public/v1/submit/delete_oauth_providers";
            const bodyWithType = {
                parameters,
                organizationId: organizationId ?? session?.organizationId ?? this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_DELETE_OAUTH_PROVIDERS",
            };
            const stringifiedBody = JSON.stringify(bodyWithType);
            const stamp = await activeStamper.stamp(stringifiedBody);
            return {
                body: stringifiedBody,
                stamp: stamp,
                url: fullUrl,
            };
        };
        this.deletePolicies = async (input, stampWith) => {
            const { organizationId, timestampMs, ...rest } = input;
            const session = await this.storageManager?.getActiveSession();
            return this.activity("/public/v1/submit/delete_policies", {
                parameters: rest,
                organizationId: organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_DELETE_POLICIES",
            }, "deletePoliciesResult", stampWith);
        };
        this.stampDeletePolicies = async (input, stampWith) => {
            const activeStamper = this.getStamper(stampWith);
            if (!activeStamper) {
                return undefined;
            }
            const { organizationId, timestampMs, ...parameters } = input;
            const session = await this.storageManager?.getActiveSession();
            const fullUrl = this.config.apiBaseUrl + "/public/v1/submit/delete_policies";
            const bodyWithType = {
                parameters,
                organizationId: organizationId ?? session?.organizationId ?? this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_DELETE_POLICIES",
            };
            const stringifiedBody = JSON.stringify(bodyWithType);
            const stamp = await activeStamper.stamp(stringifiedBody);
            return {
                body: stringifiedBody,
                stamp: stamp,
                url: fullUrl,
            };
        };
        this.deletePolicy = async (input, stampWith) => {
            const { organizationId, timestampMs, ...rest } = input;
            const session = await this.storageManager?.getActiveSession();
            return this.activity("/public/v1/submit/delete_policy", {
                parameters: rest,
                organizationId: organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_DELETE_POLICY",
            }, "deletePolicyResult", stampWith);
        };
        this.stampDeletePolicy = async (input, stampWith) => {
            const activeStamper = this.getStamper(stampWith);
            if (!activeStamper) {
                return undefined;
            }
            const { organizationId, timestampMs, ...parameters } = input;
            const session = await this.storageManager?.getActiveSession();
            const fullUrl = this.config.apiBaseUrl + "/public/v1/submit/delete_policy";
            const bodyWithType = {
                parameters,
                organizationId: organizationId ?? session?.organizationId ?? this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_DELETE_POLICY",
            };
            const stringifiedBody = JSON.stringify(bodyWithType);
            const stamp = await activeStamper.stamp(stringifiedBody);
            return {
                body: stringifiedBody,
                stamp: stamp,
                url: fullUrl,
            };
        };
        this.deletePrivateKeyTags = async (input, stampWith) => {
            const { organizationId, timestampMs, ...rest } = input;
            const session = await this.storageManager?.getActiveSession();
            return this.activity("/public/v1/submit/delete_private_key_tags", {
                parameters: rest,
                organizationId: organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_DELETE_PRIVATE_KEY_TAGS",
            }, "deletePrivateKeyTagsResult", stampWith);
        };
        this.stampDeletePrivateKeyTags = async (input, stampWith) => {
            const activeStamper = this.getStamper(stampWith);
            if (!activeStamper) {
                return undefined;
            }
            const { organizationId, timestampMs, ...parameters } = input;
            const session = await this.storageManager?.getActiveSession();
            const fullUrl = this.config.apiBaseUrl + "/public/v1/submit/delete_private_key_tags";
            const bodyWithType = {
                parameters,
                organizationId: organizationId ?? session?.organizationId ?? this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_DELETE_PRIVATE_KEY_TAGS",
            };
            const stringifiedBody = JSON.stringify(bodyWithType);
            const stamp = await activeStamper.stamp(stringifiedBody);
            return {
                body: stringifiedBody,
                stamp: stamp,
                url: fullUrl,
            };
        };
        this.deletePrivateKeys = async (input, stampWith) => {
            const { organizationId, timestampMs, ...rest } = input;
            const session = await this.storageManager?.getActiveSession();
            return this.activity("/public/v1/submit/delete_private_keys", {
                parameters: rest,
                organizationId: organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_DELETE_PRIVATE_KEYS",
            }, "deletePrivateKeysResult", stampWith);
        };
        this.stampDeletePrivateKeys = async (input, stampWith) => {
            const activeStamper = this.getStamper(stampWith);
            if (!activeStamper) {
                return undefined;
            }
            const { organizationId, timestampMs, ...parameters } = input;
            const session = await this.storageManager?.getActiveSession();
            const fullUrl = this.config.apiBaseUrl + "/public/v1/submit/delete_private_keys";
            const bodyWithType = {
                parameters,
                organizationId: organizationId ?? session?.organizationId ?? this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_DELETE_PRIVATE_KEYS",
            };
            const stringifiedBody = JSON.stringify(bodyWithType);
            const stamp = await activeStamper.stamp(stringifiedBody);
            return {
                body: stringifiedBody,
                stamp: stamp,
                url: fullUrl,
            };
        };
        this.deleteSmartContractInterface = async (input, stampWith) => {
            const { organizationId, timestampMs, ...rest } = input;
            const session = await this.storageManager?.getActiveSession();
            return this.activity("/public/v1/submit/delete_smart_contract_interface", {
                parameters: rest,
                organizationId: organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_DELETE_SMART_CONTRACT_INTERFACE",
            }, "deleteSmartContractInterfaceResult", stampWith);
        };
        this.stampDeleteSmartContractInterface = async (input, stampWith) => {
            const activeStamper = this.getStamper(stampWith);
            if (!activeStamper) {
                return undefined;
            }
            const { organizationId, timestampMs, ...parameters } = input;
            const session = await this.storageManager?.getActiveSession();
            const fullUrl = this.config.apiBaseUrl +
                "/public/v1/submit/delete_smart_contract_interface";
            const bodyWithType = {
                parameters,
                organizationId: organizationId ?? session?.organizationId ?? this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_DELETE_SMART_CONTRACT_INTERFACE",
            };
            const stringifiedBody = JSON.stringify(bodyWithType);
            const stamp = await activeStamper.stamp(stringifiedBody);
            return {
                body: stringifiedBody,
                stamp: stamp,
                url: fullUrl,
            };
        };
        this.deleteSubOrganization = async (input, stampWith) => {
            const { organizationId, timestampMs, ...rest } = input;
            const session = await this.storageManager?.getActiveSession();
            return this.activity("/public/v1/submit/delete_sub_organization", {
                parameters: rest,
                organizationId: organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_DELETE_SUB_ORGANIZATION",
            }, "deleteSubOrganizationResult", stampWith);
        };
        this.stampDeleteSubOrganization = async (input, stampWith) => {
            const activeStamper = this.getStamper(stampWith);
            if (!activeStamper) {
                return undefined;
            }
            const { organizationId, timestampMs, ...parameters } = input;
            const session = await this.storageManager?.getActiveSession();
            const fullUrl = this.config.apiBaseUrl + "/public/v1/submit/delete_sub_organization";
            const bodyWithType = {
                parameters,
                organizationId: organizationId ?? session?.organizationId ?? this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_DELETE_SUB_ORGANIZATION",
            };
            const stringifiedBody = JSON.stringify(bodyWithType);
            const stamp = await activeStamper.stamp(stringifiedBody);
            return {
                body: stringifiedBody,
                stamp: stamp,
                url: fullUrl,
            };
        };
        this.deleteUserTags = async (input, stampWith) => {
            const { organizationId, timestampMs, ...rest } = input;
            const session = await this.storageManager?.getActiveSession();
            return this.activity("/public/v1/submit/delete_user_tags", {
                parameters: rest,
                organizationId: organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_DELETE_USER_TAGS",
            }, "deleteUserTagsResult", stampWith);
        };
        this.stampDeleteUserTags = async (input, stampWith) => {
            const activeStamper = this.getStamper(stampWith);
            if (!activeStamper) {
                return undefined;
            }
            const { organizationId, timestampMs, ...parameters } = input;
            const session = await this.storageManager?.getActiveSession();
            const fullUrl = this.config.apiBaseUrl + "/public/v1/submit/delete_user_tags";
            const bodyWithType = {
                parameters,
                organizationId: organizationId ?? session?.organizationId ?? this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_DELETE_USER_TAGS",
            };
            const stringifiedBody = JSON.stringify(bodyWithType);
            const stamp = await activeStamper.stamp(stringifiedBody);
            return {
                body: stringifiedBody,
                stamp: stamp,
                url: fullUrl,
            };
        };
        this.deleteUsers = async (input, stampWith) => {
            const { organizationId, timestampMs, ...rest } = input;
            const session = await this.storageManager?.getActiveSession();
            return this.activity("/public/v1/submit/delete_users", {
                parameters: rest,
                organizationId: organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_DELETE_USERS",
            }, "deleteUsersResult", stampWith);
        };
        this.stampDeleteUsers = async (input, stampWith) => {
            const activeStamper = this.getStamper(stampWith);
            if (!activeStamper) {
                return undefined;
            }
            const { organizationId, timestampMs, ...parameters } = input;
            const session = await this.storageManager?.getActiveSession();
            const fullUrl = this.config.apiBaseUrl + "/public/v1/submit/delete_users";
            const bodyWithType = {
                parameters,
                organizationId: organizationId ?? session?.organizationId ?? this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_DELETE_USERS",
            };
            const stringifiedBody = JSON.stringify(bodyWithType);
            const stamp = await activeStamper.stamp(stringifiedBody);
            return {
                body: stringifiedBody,
                stamp: stamp,
                url: fullUrl,
            };
        };
        this.deleteWalletAccounts = async (input, stampWith) => {
            const { organizationId, timestampMs, ...rest } = input;
            const session = await this.storageManager?.getActiveSession();
            return this.activity("/public/v1/submit/delete_wallet_accounts", {
                parameters: rest,
                organizationId: organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_DELETE_WALLET_ACCOUNTS",
            }, "deleteWalletAccountsResult", stampWith);
        };
        this.stampDeleteWalletAccounts = async (input, stampWith) => {
            const activeStamper = this.getStamper(stampWith);
            if (!activeStamper) {
                return undefined;
            }
            const { organizationId, timestampMs, ...parameters } = input;
            const session = await this.storageManager?.getActiveSession();
            const fullUrl = this.config.apiBaseUrl + "/public/v1/submit/delete_wallet_accounts";
            const bodyWithType = {
                parameters,
                organizationId: organizationId ?? session?.organizationId ?? this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_DELETE_WALLET_ACCOUNTS",
            };
            const stringifiedBody = JSON.stringify(bodyWithType);
            const stamp = await activeStamper.stamp(stringifiedBody);
            return {
                body: stringifiedBody,
                stamp: stamp,
                url: fullUrl,
            };
        };
        this.deleteWallets = async (input, stampWith) => {
            const { organizationId, timestampMs, ...rest } = input;
            const session = await this.storageManager?.getActiveSession();
            return this.activity("/public/v1/submit/delete_wallets", {
                parameters: rest,
                organizationId: organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_DELETE_WALLETS",
            }, "deleteWalletsResult", stampWith);
        };
        this.stampDeleteWallets = async (input, stampWith) => {
            const activeStamper = this.getStamper(stampWith);
            if (!activeStamper) {
                return undefined;
            }
            const { organizationId, timestampMs, ...parameters } = input;
            const session = await this.storageManager?.getActiveSession();
            const fullUrl = this.config.apiBaseUrl + "/public/v1/submit/delete_wallets";
            const bodyWithType = {
                parameters,
                organizationId: organizationId ?? session?.organizationId ?? this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_DELETE_WALLETS",
            };
            const stringifiedBody = JSON.stringify(bodyWithType);
            const stamp = await activeStamper.stamp(stringifiedBody);
            return {
                body: stringifiedBody,
                stamp: stamp,
                url: fullUrl,
            };
        };
        this.emailAuth = async (input, stampWith) => {
            const { organizationId, timestampMs, ...rest } = input;
            const session = await this.storageManager?.getActiveSession();
            return this.activity("/public/v1/submit/email_auth", {
                parameters: rest,
                organizationId: organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_EMAIL_AUTH_V3",
            }, "emailAuthResult", stampWith);
        };
        this.stampEmailAuth = async (input, stampWith) => {
            const activeStamper = this.getStamper(stampWith);
            if (!activeStamper) {
                return undefined;
            }
            const { organizationId, timestampMs, ...parameters } = input;
            const session = await this.storageManager?.getActiveSession();
            const fullUrl = this.config.apiBaseUrl + "/public/v1/submit/email_auth";
            const bodyWithType = {
                parameters,
                organizationId: organizationId ?? session?.organizationId ?? this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_EMAIL_AUTH_V3",
            };
            const stringifiedBody = JSON.stringify(bodyWithType);
            const stamp = await activeStamper.stamp(stringifiedBody);
            return {
                body: stringifiedBody,
                stamp: stamp,
                url: fullUrl,
            };
        };
        this.ethSendTransaction = async (input, stampWith) => {
            const { organizationId, timestampMs, ...rest } = input;
            const session = await this.storageManager?.getActiveSession();
            return this.activity("/public/v1/submit/eth_send_transaction", {
                parameters: rest,
                organizationId: organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_ETH_SEND_TRANSACTION",
            }, "ethSendTransactionResult", stampWith);
        };
        this.stampEthSendTransaction = async (input, stampWith) => {
            const activeStamper = this.getStamper(stampWith);
            if (!activeStamper) {
                return undefined;
            }
            const { organizationId, timestampMs, ...parameters } = input;
            const session = await this.storageManager?.getActiveSession();
            const fullUrl = this.config.apiBaseUrl + "/public/v1/submit/eth_send_transaction";
            const bodyWithType = {
                parameters,
                organizationId: organizationId ?? session?.organizationId ?? this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_ETH_SEND_TRANSACTION",
            };
            const stringifiedBody = JSON.stringify(bodyWithType);
            const stamp = await activeStamper.stamp(stringifiedBody);
            return {
                body: stringifiedBody,
                stamp: stamp,
                url: fullUrl,
            };
        };
        this.exportPrivateKey = async (input, stampWith) => {
            const { organizationId, timestampMs, ...rest } = input;
            const session = await this.storageManager?.getActiveSession();
            return this.activity("/public/v1/submit/export_private_key", {
                parameters: rest,
                organizationId: organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_EXPORT_PRIVATE_KEY",
            }, "exportPrivateKeyResult", stampWith);
        };
        this.stampExportPrivateKey = async (input, stampWith) => {
            const activeStamper = this.getStamper(stampWith);
            if (!activeStamper) {
                return undefined;
            }
            const { organizationId, timestampMs, ...parameters } = input;
            const session = await this.storageManager?.getActiveSession();
            const fullUrl = this.config.apiBaseUrl + "/public/v1/submit/export_private_key";
            const bodyWithType = {
                parameters,
                organizationId: organizationId ?? session?.organizationId ?? this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_EXPORT_PRIVATE_KEY",
            };
            const stringifiedBody = JSON.stringify(bodyWithType);
            const stamp = await activeStamper.stamp(stringifiedBody);
            return {
                body: stringifiedBody,
                stamp: stamp,
                url: fullUrl,
            };
        };
        this.exportWallet = async (input, stampWith) => {
            const { organizationId, timestampMs, ...rest } = input;
            const session = await this.storageManager?.getActiveSession();
            return this.activity("/public/v1/submit/export_wallet", {
                parameters: rest,
                organizationId: organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_EXPORT_WALLET",
            }, "exportWalletResult", stampWith);
        };
        this.stampExportWallet = async (input, stampWith) => {
            const activeStamper = this.getStamper(stampWith);
            if (!activeStamper) {
                return undefined;
            }
            const { organizationId, timestampMs, ...parameters } = input;
            const session = await this.storageManager?.getActiveSession();
            const fullUrl = this.config.apiBaseUrl + "/public/v1/submit/export_wallet";
            const bodyWithType = {
                parameters,
                organizationId: organizationId ?? session?.organizationId ?? this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_EXPORT_WALLET",
            };
            const stringifiedBody = JSON.stringify(bodyWithType);
            const stamp = await activeStamper.stamp(stringifiedBody);
            return {
                body: stringifiedBody,
                stamp: stamp,
                url: fullUrl,
            };
        };
        this.exportWalletAccount = async (input, stampWith) => {
            const { organizationId, timestampMs, ...rest } = input;
            const session = await this.storageManager?.getActiveSession();
            return this.activity("/public/v1/submit/export_wallet_account", {
                parameters: rest,
                organizationId: organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_EXPORT_WALLET_ACCOUNT",
            }, "exportWalletAccountResult", stampWith);
        };
        this.stampExportWalletAccount = async (input, stampWith) => {
            const activeStamper = this.getStamper(stampWith);
            if (!activeStamper) {
                return undefined;
            }
            const { organizationId, timestampMs, ...parameters } = input;
            const session = await this.storageManager?.getActiveSession();
            const fullUrl = this.config.apiBaseUrl + "/public/v1/submit/export_wallet_account";
            const bodyWithType = {
                parameters,
                organizationId: organizationId ?? session?.organizationId ?? this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_EXPORT_WALLET_ACCOUNT",
            };
            const stringifiedBody = JSON.stringify(bodyWithType);
            const stamp = await activeStamper.stamp(stringifiedBody);
            return {
                body: stringifiedBody,
                stamp: stamp,
                url: fullUrl,
            };
        };
        this.importPrivateKey = async (input, stampWith) => {
            const { organizationId, timestampMs, ...rest } = input;
            const session = await this.storageManager?.getActiveSession();
            return this.activity("/public/v1/submit/import_private_key", {
                parameters: rest,
                organizationId: organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_IMPORT_PRIVATE_KEY",
            }, "importPrivateKeyResult", stampWith);
        };
        this.stampImportPrivateKey = async (input, stampWith) => {
            const activeStamper = this.getStamper(stampWith);
            if (!activeStamper) {
                return undefined;
            }
            const { organizationId, timestampMs, ...parameters } = input;
            const session = await this.storageManager?.getActiveSession();
            const fullUrl = this.config.apiBaseUrl + "/public/v1/submit/import_private_key";
            const bodyWithType = {
                parameters,
                organizationId: organizationId ?? session?.organizationId ?? this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_IMPORT_PRIVATE_KEY",
            };
            const stringifiedBody = JSON.stringify(bodyWithType);
            const stamp = await activeStamper.stamp(stringifiedBody);
            return {
                body: stringifiedBody,
                stamp: stamp,
                url: fullUrl,
            };
        };
        this.importWallet = async (input, stampWith) => {
            const { organizationId, timestampMs, ...rest } = input;
            const session = await this.storageManager?.getActiveSession();
            return this.activity("/public/v1/submit/import_wallet", {
                parameters: rest,
                organizationId: organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_IMPORT_WALLET",
            }, "importWalletResult", stampWith);
        };
        this.stampImportWallet = async (input, stampWith) => {
            const activeStamper = this.getStamper(stampWith);
            if (!activeStamper) {
                return undefined;
            }
            const { organizationId, timestampMs, ...parameters } = input;
            const session = await this.storageManager?.getActiveSession();
            const fullUrl = this.config.apiBaseUrl + "/public/v1/submit/import_wallet";
            const bodyWithType = {
                parameters,
                organizationId: organizationId ?? session?.organizationId ?? this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_IMPORT_WALLET",
            };
            const stringifiedBody = JSON.stringify(bodyWithType);
            const stamp = await activeStamper.stamp(stringifiedBody);
            return {
                body: stringifiedBody,
                stamp: stamp,
                url: fullUrl,
            };
        };
        this.initFiatOnRamp = async (input, stampWith) => {
            const { organizationId, timestampMs, ...rest } = input;
            const session = await this.storageManager?.getActiveSession();
            return this.activity("/public/v1/submit/init_fiat_on_ramp", {
                parameters: rest,
                organizationId: organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_INIT_FIAT_ON_RAMP",
            }, "initFiatOnRampResult", stampWith);
        };
        this.stampInitFiatOnRamp = async (input, stampWith) => {
            const activeStamper = this.getStamper(stampWith);
            if (!activeStamper) {
                return undefined;
            }
            const { organizationId, timestampMs, ...parameters } = input;
            const session = await this.storageManager?.getActiveSession();
            const fullUrl = this.config.apiBaseUrl + "/public/v1/submit/init_fiat_on_ramp";
            const bodyWithType = {
                parameters,
                organizationId: organizationId ?? session?.organizationId ?? this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_INIT_FIAT_ON_RAMP",
            };
            const stringifiedBody = JSON.stringify(bodyWithType);
            const stamp = await activeStamper.stamp(stringifiedBody);
            return {
                body: stringifiedBody,
                stamp: stamp,
                url: fullUrl,
            };
        };
        this.initImportPrivateKey = async (input, stampWith) => {
            const { organizationId, timestampMs, ...rest } = input;
            const session = await this.storageManager?.getActiveSession();
            return this.activity("/public/v1/submit/init_import_private_key", {
                parameters: rest,
                organizationId: organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_INIT_IMPORT_PRIVATE_KEY",
            }, "initImportPrivateKeyResult", stampWith);
        };
        this.stampInitImportPrivateKey = async (input, stampWith) => {
            const activeStamper = this.getStamper(stampWith);
            if (!activeStamper) {
                return undefined;
            }
            const { organizationId, timestampMs, ...parameters } = input;
            const session = await this.storageManager?.getActiveSession();
            const fullUrl = this.config.apiBaseUrl + "/public/v1/submit/init_import_private_key";
            const bodyWithType = {
                parameters,
                organizationId: organizationId ?? session?.organizationId ?? this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_INIT_IMPORT_PRIVATE_KEY",
            };
            const stringifiedBody = JSON.stringify(bodyWithType);
            const stamp = await activeStamper.stamp(stringifiedBody);
            return {
                body: stringifiedBody,
                stamp: stamp,
                url: fullUrl,
            };
        };
        this.initImportWallet = async (input, stampWith) => {
            const { organizationId, timestampMs, ...rest } = input;
            const session = await this.storageManager?.getActiveSession();
            return this.activity("/public/v1/submit/init_import_wallet", {
                parameters: rest,
                organizationId: organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_INIT_IMPORT_WALLET",
            }, "initImportWalletResult", stampWith);
        };
        this.stampInitImportWallet = async (input, stampWith) => {
            const activeStamper = this.getStamper(stampWith);
            if (!activeStamper) {
                return undefined;
            }
            const { organizationId, timestampMs, ...parameters } = input;
            const session = await this.storageManager?.getActiveSession();
            const fullUrl = this.config.apiBaseUrl + "/public/v1/submit/init_import_wallet";
            const bodyWithType = {
                parameters,
                organizationId: organizationId ?? session?.organizationId ?? this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_INIT_IMPORT_WALLET",
            };
            const stringifiedBody = JSON.stringify(bodyWithType);
            const stamp = await activeStamper.stamp(stringifiedBody);
            return {
                body: stringifiedBody,
                stamp: stamp,
                url: fullUrl,
            };
        };
        this.initOtp = async (input, stampWith) => {
            const { organizationId, timestampMs, ...rest } = input;
            const session = await this.storageManager?.getActiveSession();
            return this.activity("/public/v1/submit/init_otp", {
                parameters: rest,
                organizationId: organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_INIT_OTP_V2",
            }, "initOtpResult", stampWith);
        };
        this.stampInitOtp = async (input, stampWith) => {
            const activeStamper = this.getStamper(stampWith);
            if (!activeStamper) {
                return undefined;
            }
            const { organizationId, timestampMs, ...parameters } = input;
            const session = await this.storageManager?.getActiveSession();
            const fullUrl = this.config.apiBaseUrl + "/public/v1/submit/init_otp";
            const bodyWithType = {
                parameters,
                organizationId: organizationId ?? session?.organizationId ?? this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_INIT_OTP_V2",
            };
            const stringifiedBody = JSON.stringify(bodyWithType);
            const stamp = await activeStamper.stamp(stringifiedBody);
            return {
                body: stringifiedBody,
                stamp: stamp,
                url: fullUrl,
            };
        };
        this.initOtpAuth = async (input, stampWith) => {
            const { organizationId, timestampMs, ...rest } = input;
            const session = await this.storageManager?.getActiveSession();
            return this.activity("/public/v1/submit/init_otp_auth", {
                parameters: rest,
                organizationId: organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_INIT_OTP_AUTH_V3",
            }, "initOtpAuthResultV2", stampWith);
        };
        this.stampInitOtpAuth = async (input, stampWith) => {
            const activeStamper = this.getStamper(stampWith);
            if (!activeStamper) {
                return undefined;
            }
            const { organizationId, timestampMs, ...parameters } = input;
            const session = await this.storageManager?.getActiveSession();
            const fullUrl = this.config.apiBaseUrl + "/public/v1/submit/init_otp_auth";
            const bodyWithType = {
                parameters,
                organizationId: organizationId ?? session?.organizationId ?? this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_INIT_OTP_AUTH_V3",
            };
            const stringifiedBody = JSON.stringify(bodyWithType);
            const stamp = await activeStamper.stamp(stringifiedBody);
            return {
                body: stringifiedBody,
                stamp: stamp,
                url: fullUrl,
            };
        };
        this.initUserEmailRecovery = async (input, stampWith) => {
            const { organizationId, timestampMs, ...rest } = input;
            const session = await this.storageManager?.getActiveSession();
            return this.activity("/public/v1/submit/init_user_email_recovery", {
                parameters: rest,
                organizationId: organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_INIT_USER_EMAIL_RECOVERY_V2",
            }, "initUserEmailRecoveryResult", stampWith);
        };
        this.stampInitUserEmailRecovery = async (input, stampWith) => {
            const activeStamper = this.getStamper(stampWith);
            if (!activeStamper) {
                return undefined;
            }
            const { organizationId, timestampMs, ...parameters } = input;
            const session = await this.storageManager?.getActiveSession();
            const fullUrl = this.config.apiBaseUrl + "/public/v1/submit/init_user_email_recovery";
            const bodyWithType = {
                parameters,
                organizationId: organizationId ?? session?.organizationId ?? this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_INIT_USER_EMAIL_RECOVERY_V2",
            };
            const stringifiedBody = JSON.stringify(bodyWithType);
            const stamp = await activeStamper.stamp(stringifiedBody);
            return {
                body: stringifiedBody,
                stamp: stamp,
                url: fullUrl,
            };
        };
        this.oauth = async (input, stampWith) => {
            const { organizationId, timestampMs, ...rest } = input;
            const session = await this.storageManager?.getActiveSession();
            return this.activity("/public/v1/submit/oauth", {
                parameters: rest,
                organizationId: organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_OAUTH",
            }, "oauthResult", stampWith);
        };
        this.stampOauth = async (input, stampWith) => {
            const activeStamper = this.getStamper(stampWith);
            if (!activeStamper) {
                return undefined;
            }
            const { organizationId, timestampMs, ...parameters } = input;
            const session = await this.storageManager?.getActiveSession();
            const fullUrl = this.config.apiBaseUrl + "/public/v1/submit/oauth";
            const bodyWithType = {
                parameters,
                organizationId: organizationId ?? session?.organizationId ?? this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_OAUTH",
            };
            const stringifiedBody = JSON.stringify(bodyWithType);
            const stamp = await activeStamper.stamp(stringifiedBody);
            return {
                body: stringifiedBody,
                stamp: stamp,
                url: fullUrl,
            };
        };
        this.oauth2Authenticate = async (input, stampWith) => {
            const { organizationId, timestampMs, ...rest } = input;
            const session = await this.storageManager?.getActiveSession();
            return this.activity("/public/v1/submit/oauth2_authenticate", {
                parameters: rest,
                organizationId: organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_OAUTH2AUTHENTICATE",
            }, "oauth2AuthenticateResult", stampWith);
        };
        this.stampOauth2Authenticate = async (input, stampWith) => {
            const activeStamper = this.getStamper(stampWith);
            if (!activeStamper) {
                return undefined;
            }
            const { organizationId, timestampMs, ...parameters } = input;
            const session = await this.storageManager?.getActiveSession();
            const fullUrl = this.config.apiBaseUrl + "/public/v1/submit/oauth2_authenticate";
            const bodyWithType = {
                parameters,
                organizationId: organizationId ?? session?.organizationId ?? this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_OAUTH2AUTHENTICATE",
            };
            const stringifiedBody = JSON.stringify(bodyWithType);
            const stamp = await activeStamper.stamp(stringifiedBody);
            return {
                body: stringifiedBody,
                stamp: stamp,
                url: fullUrl,
            };
        };
        this.oauthLogin = async (input, stampWith) => {
            const { organizationId, timestampMs, ...rest } = input;
            const session = await this.storageManager?.getActiveSession();
            return this.activity("/public/v1/submit/oauth_login", {
                parameters: rest,
                organizationId: organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_OAUTH_LOGIN",
            }, "oauthLoginResult", stampWith);
        };
        this.stampOauthLogin = async (input, stampWith) => {
            const activeStamper = this.getStamper(stampWith);
            if (!activeStamper) {
                return undefined;
            }
            const { organizationId, timestampMs, ...parameters } = input;
            const session = await this.storageManager?.getActiveSession();
            const fullUrl = this.config.apiBaseUrl + "/public/v1/submit/oauth_login";
            const bodyWithType = {
                parameters,
                organizationId: organizationId ?? session?.organizationId ?? this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_OAUTH_LOGIN",
            };
            const stringifiedBody = JSON.stringify(bodyWithType);
            const stamp = await activeStamper.stamp(stringifiedBody);
            return {
                body: stringifiedBody,
                stamp: stamp,
                url: fullUrl,
            };
        };
        this.otpAuth = async (input, stampWith) => {
            const { organizationId, timestampMs, ...rest } = input;
            const session = await this.storageManager?.getActiveSession();
            return this.activity("/public/v1/submit/otp_auth", {
                parameters: rest,
                organizationId: organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_OTP_AUTH",
            }, "otpAuthResult", stampWith);
        };
        this.stampOtpAuth = async (input, stampWith) => {
            const activeStamper = this.getStamper(stampWith);
            if (!activeStamper) {
                return undefined;
            }
            const { organizationId, timestampMs, ...parameters } = input;
            const session = await this.storageManager?.getActiveSession();
            const fullUrl = this.config.apiBaseUrl + "/public/v1/submit/otp_auth";
            const bodyWithType = {
                parameters,
                organizationId: organizationId ?? session?.organizationId ?? this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_OTP_AUTH",
            };
            const stringifiedBody = JSON.stringify(bodyWithType);
            const stamp = await activeStamper.stamp(stringifiedBody);
            return {
                body: stringifiedBody,
                stamp: stamp,
                url: fullUrl,
            };
        };
        this.otpLogin = async (input, stampWith) => {
            const { organizationId, timestampMs, ...rest } = input;
            const session = await this.storageManager?.getActiveSession();
            return this.activity("/public/v1/submit/otp_login", {
                parameters: rest,
                organizationId: organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_OTP_LOGIN",
            }, "otpLoginResult", stampWith);
        };
        this.stampOtpLogin = async (input, stampWith) => {
            const activeStamper = this.getStamper(stampWith);
            if (!activeStamper) {
                return undefined;
            }
            const { organizationId, timestampMs, ...parameters } = input;
            const session = await this.storageManager?.getActiveSession();
            const fullUrl = this.config.apiBaseUrl + "/public/v1/submit/otp_login";
            const bodyWithType = {
                parameters,
                organizationId: organizationId ?? session?.organizationId ?? this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_OTP_LOGIN",
            };
            const stringifiedBody = JSON.stringify(bodyWithType);
            const stamp = await activeStamper.stamp(stringifiedBody);
            return {
                body: stringifiedBody,
                stamp: stamp,
                url: fullUrl,
            };
        };
        this.recoverUser = async (input, stampWith) => {
            const { organizationId, timestampMs, ...rest } = input;
            const session = await this.storageManager?.getActiveSession();
            return this.activity("/public/v1/submit/recover_user", {
                parameters: rest,
                organizationId: organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_RECOVER_USER",
            }, "recoverUserResult", stampWith);
        };
        this.stampRecoverUser = async (input, stampWith) => {
            const activeStamper = this.getStamper(stampWith);
            if (!activeStamper) {
                return undefined;
            }
            const { organizationId, timestampMs, ...parameters } = input;
            const session = await this.storageManager?.getActiveSession();
            const fullUrl = this.config.apiBaseUrl + "/public/v1/submit/recover_user";
            const bodyWithType = {
                parameters,
                organizationId: organizationId ?? session?.organizationId ?? this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_RECOVER_USER",
            };
            const stringifiedBody = JSON.stringify(bodyWithType);
            const stamp = await activeStamper.stamp(stringifiedBody);
            return {
                body: stringifiedBody,
                stamp: stamp,
                url: fullUrl,
            };
        };
        this.rejectActivity = async (input, stampWith) => {
            const { organizationId, timestampMs, ...rest } = input;
            const session = await this.storageManager?.getActiveSession();
            return this.activityDecision("/public/v1/submit/reject_activity", {
                parameters: rest,
                organizationId: organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_REJECT_ACTIVITY",
            }, stampWith);
        };
        this.stampRejectActivity = async (input, stampWith) => {
            const activeStamper = this.getStamper(stampWith);
            if (!activeStamper) {
                return undefined;
            }
            const { organizationId, timestampMs, ...parameters } = input;
            const session = await this.storageManager?.getActiveSession();
            const fullUrl = this.config.apiBaseUrl + "/public/v1/submit/reject_activity";
            const bodyWithType = {
                parameters,
                organizationId: organizationId ?? session?.organizationId ?? this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_REJECT_ACTIVITY",
            };
            const stringifiedBody = JSON.stringify(bodyWithType);
            const stamp = await activeStamper.stamp(stringifiedBody);
            return {
                body: stringifiedBody,
                stamp: stamp,
                url: fullUrl,
            };
        };
        this.removeOrganizationFeature = async (input, stampWith) => {
            const { organizationId, timestampMs, ...rest } = input;
            const session = await this.storageManager?.getActiveSession();
            return this.activity("/public/v1/submit/remove_organization_feature", {
                parameters: rest,
                organizationId: organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_REMOVE_ORGANIZATION_FEATURE",
            }, "removeOrganizationFeatureResult", stampWith);
        };
        this.stampRemoveOrganizationFeature = async (input, stampWith) => {
            const activeStamper = this.getStamper(stampWith);
            if (!activeStamper) {
                return undefined;
            }
            const { organizationId, timestampMs, ...parameters } = input;
            const session = await this.storageManager?.getActiveSession();
            const fullUrl = this.config.apiBaseUrl + "/public/v1/submit/remove_organization_feature";
            const bodyWithType = {
                parameters,
                organizationId: organizationId ?? session?.organizationId ?? this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_REMOVE_ORGANIZATION_FEATURE",
            };
            const stringifiedBody = JSON.stringify(bodyWithType);
            const stamp = await activeStamper.stamp(stringifiedBody);
            return {
                body: stringifiedBody,
                stamp: stamp,
                url: fullUrl,
            };
        };
        this.setOrganizationFeature = async (input, stampWith) => {
            const { organizationId, timestampMs, ...rest } = input;
            const session = await this.storageManager?.getActiveSession();
            return this.activity("/public/v1/submit/set_organization_feature", {
                parameters: rest,
                organizationId: organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_SET_ORGANIZATION_FEATURE",
            }, "setOrganizationFeatureResult", stampWith);
        };
        this.stampSetOrganizationFeature = async (input, stampWith) => {
            const activeStamper = this.getStamper(stampWith);
            if (!activeStamper) {
                return undefined;
            }
            const { organizationId, timestampMs, ...parameters } = input;
            const session = await this.storageManager?.getActiveSession();
            const fullUrl = this.config.apiBaseUrl + "/public/v1/submit/set_organization_feature";
            const bodyWithType = {
                parameters,
                organizationId: organizationId ?? session?.organizationId ?? this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_SET_ORGANIZATION_FEATURE",
            };
            const stringifiedBody = JSON.stringify(bodyWithType);
            const stamp = await activeStamper.stamp(stringifiedBody);
            return {
                body: stringifiedBody,
                stamp: stamp,
                url: fullUrl,
            };
        };
        this.signRawPayload = async (input, stampWith) => {
            const { organizationId, timestampMs, ...rest } = input;
            const session = await this.storageManager?.getActiveSession();
            return this.activity("/public/v1/submit/sign_raw_payload", {
                parameters: rest,
                organizationId: organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_SIGN_RAW_PAYLOAD_V2",
            }, "signRawPayloadResult", stampWith);
        };
        this.stampSignRawPayload = async (input, stampWith) => {
            const activeStamper = this.getStamper(stampWith);
            if (!activeStamper) {
                return undefined;
            }
            const { organizationId, timestampMs, ...parameters } = input;
            const session = await this.storageManager?.getActiveSession();
            const fullUrl = this.config.apiBaseUrl + "/public/v1/submit/sign_raw_payload";
            const bodyWithType = {
                parameters,
                organizationId: organizationId ?? session?.organizationId ?? this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_SIGN_RAW_PAYLOAD_V2",
            };
            const stringifiedBody = JSON.stringify(bodyWithType);
            const stamp = await activeStamper.stamp(stringifiedBody);
            return {
                body: stringifiedBody,
                stamp: stamp,
                url: fullUrl,
            };
        };
        this.signRawPayloads = async (input, stampWith) => {
            const { organizationId, timestampMs, ...rest } = input;
            const session = await this.storageManager?.getActiveSession();
            return this.activity("/public/v1/submit/sign_raw_payloads", {
                parameters: rest,
                organizationId: organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_SIGN_RAW_PAYLOADS",
            }, "signRawPayloadsResult", stampWith);
        };
        this.stampSignRawPayloads = async (input, stampWith) => {
            const activeStamper = this.getStamper(stampWith);
            if (!activeStamper) {
                return undefined;
            }
            const { organizationId, timestampMs, ...parameters } = input;
            const session = await this.storageManager?.getActiveSession();
            const fullUrl = this.config.apiBaseUrl + "/public/v1/submit/sign_raw_payloads";
            const bodyWithType = {
                parameters,
                organizationId: organizationId ?? session?.organizationId ?? this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_SIGN_RAW_PAYLOADS",
            };
            const stringifiedBody = JSON.stringify(bodyWithType);
            const stamp = await activeStamper.stamp(stringifiedBody);
            return {
                body: stringifiedBody,
                stamp: stamp,
                url: fullUrl,
            };
        };
        this.signTransaction = async (input, stampWith) => {
            const { organizationId, timestampMs, ...rest } = input;
            const session = await this.storageManager?.getActiveSession();
            return this.activity("/public/v1/submit/sign_transaction", {
                parameters: rest,
                organizationId: organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_SIGN_TRANSACTION_V2",
            }, "signTransactionResult", stampWith);
        };
        this.stampSignTransaction = async (input, stampWith) => {
            const activeStamper = this.getStamper(stampWith);
            if (!activeStamper) {
                return undefined;
            }
            const { organizationId, timestampMs, ...parameters } = input;
            const session = await this.storageManager?.getActiveSession();
            const fullUrl = this.config.apiBaseUrl + "/public/v1/submit/sign_transaction";
            const bodyWithType = {
                parameters,
                organizationId: organizationId ?? session?.organizationId ?? this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_SIGN_TRANSACTION_V2",
            };
            const stringifiedBody = JSON.stringify(bodyWithType);
            const stamp = await activeStamper.stamp(stringifiedBody);
            return {
                body: stringifiedBody,
                stamp: stamp,
                url: fullUrl,
            };
        };
        this.solSendTransaction = async (input, stampWith) => {
            const { organizationId, timestampMs, ...rest } = input;
            const session = await this.storageManager?.getActiveSession();
            return this.activity("/public/v1/submit/sol_send_transaction", {
                parameters: rest,
                organizationId: organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_SOL_SEND_TRANSACTION",
            }, "solSendTransactionResult", stampWith);
        };
        this.stampSolSendTransaction = async (input, stampWith) => {
            const activeStamper = this.getStamper(stampWith);
            if (!activeStamper) {
                return undefined;
            }
            const { organizationId, timestampMs, ...parameters } = input;
            const session = await this.storageManager?.getActiveSession();
            const fullUrl = this.config.apiBaseUrl + "/public/v1/submit/sol_send_transaction";
            const bodyWithType = {
                parameters,
                organizationId: organizationId ?? session?.organizationId ?? this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_SOL_SEND_TRANSACTION",
            };
            const stringifiedBody = JSON.stringify(bodyWithType);
            const stamp = await activeStamper.stamp(stringifiedBody);
            return {
                body: stringifiedBody,
                stamp: stamp,
                url: fullUrl,
            };
        };
        this.stampLogin = async (input, stampWith) => {
            const { organizationId, timestampMs, ...rest } = input;
            const session = await this.storageManager?.getActiveSession();
            return this.activity("/public/v1/submit/stamp_login", {
                parameters: rest,
                organizationId: organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_STAMP_LOGIN",
            }, "stampLoginResult", stampWith);
        };
        this.stampStampLogin = async (input, stampWith) => {
            const activeStamper = this.getStamper(stampWith);
            if (!activeStamper) {
                return undefined;
            }
            const { organizationId, timestampMs, ...parameters } = input;
            const session = await this.storageManager?.getActiveSession();
            const fullUrl = this.config.apiBaseUrl + "/public/v1/submit/stamp_login";
            const bodyWithType = {
                parameters,
                organizationId: organizationId ?? session?.organizationId ?? this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_STAMP_LOGIN",
            };
            const stringifiedBody = JSON.stringify(bodyWithType);
            const stamp = await activeStamper.stamp(stringifiedBody);
            return {
                body: stringifiedBody,
                stamp: stamp,
                url: fullUrl,
            };
        };
        this.updateFiatOnRampCredential = async (input, stampWith) => {
            const { organizationId, timestampMs, ...rest } = input;
            const session = await this.storageManager?.getActiveSession();
            return this.activity("/public/v1/submit/update_fiat_on_ramp_credential", {
                parameters: rest,
                organizationId: organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_UPDATE_FIAT_ON_RAMP_CREDENTIAL",
            }, "updateFiatOnRampCredentialResult", stampWith);
        };
        this.stampUpdateFiatOnRampCredential = async (input, stampWith) => {
            const activeStamper = this.getStamper(stampWith);
            if (!activeStamper) {
                return undefined;
            }
            const { organizationId, timestampMs, ...parameters } = input;
            const session = await this.storageManager?.getActiveSession();
            const fullUrl = this.config.apiBaseUrl +
                "/public/v1/submit/update_fiat_on_ramp_credential";
            const bodyWithType = {
                parameters,
                organizationId: organizationId ?? session?.organizationId ?? this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_UPDATE_FIAT_ON_RAMP_CREDENTIAL",
            };
            const stringifiedBody = JSON.stringify(bodyWithType);
            const stamp = await activeStamper.stamp(stringifiedBody);
            return {
                body: stringifiedBody,
                stamp: stamp,
                url: fullUrl,
            };
        };
        this.updateOauth2Credential = async (input, stampWith) => {
            const { organizationId, timestampMs, ...rest } = input;
            const session = await this.storageManager?.getActiveSession();
            return this.activity("/public/v1/submit/update_oauth2_credential", {
                parameters: rest,
                organizationId: organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_UPDATE_OAUTH2CREDENTIAL",
            }, "updateOauth2CredentialResult", stampWith);
        };
        this.stampUpdateOauth2Credential = async (input, stampWith) => {
            const activeStamper = this.getStamper(stampWith);
            if (!activeStamper) {
                return undefined;
            }
            const { organizationId, timestampMs, ...parameters } = input;
            const session = await this.storageManager?.getActiveSession();
            const fullUrl = this.config.apiBaseUrl + "/public/v1/submit/update_oauth2_credential";
            const bodyWithType = {
                parameters,
                organizationId: organizationId ?? session?.organizationId ?? this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_UPDATE_OAUTH2CREDENTIAL",
            };
            const stringifiedBody = JSON.stringify(bodyWithType);
            const stamp = await activeStamper.stamp(stringifiedBody);
            return {
                body: stringifiedBody,
                stamp: stamp,
                url: fullUrl,
            };
        };
        this.updatePolicy = async (input, stampWith) => {
            const { organizationId, timestampMs, ...rest } = input;
            const session = await this.storageManager?.getActiveSession();
            return this.activity("/public/v1/submit/update_policy", {
                parameters: rest,
                organizationId: organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_UPDATE_POLICY_V2",
            }, "updatePolicyResultV2", stampWith);
        };
        this.stampUpdatePolicy = async (input, stampWith) => {
            const activeStamper = this.getStamper(stampWith);
            if (!activeStamper) {
                return undefined;
            }
            const { organizationId, timestampMs, ...parameters } = input;
            const session = await this.storageManager?.getActiveSession();
            const fullUrl = this.config.apiBaseUrl + "/public/v1/submit/update_policy";
            const bodyWithType = {
                parameters,
                organizationId: organizationId ?? session?.organizationId ?? this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_UPDATE_POLICY_V2",
            };
            const stringifiedBody = JSON.stringify(bodyWithType);
            const stamp = await activeStamper.stamp(stringifiedBody);
            return {
                body: stringifiedBody,
                stamp: stamp,
                url: fullUrl,
            };
        };
        this.updatePrivateKeyTag = async (input, stampWith) => {
            const { organizationId, timestampMs, ...rest } = input;
            const session = await this.storageManager?.getActiveSession();
            return this.activity("/public/v1/submit/update_private_key_tag", {
                parameters: rest,
                organizationId: organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_UPDATE_PRIVATE_KEY_TAG",
            }, "updatePrivateKeyTagResult", stampWith);
        };
        this.stampUpdatePrivateKeyTag = async (input, stampWith) => {
            const activeStamper = this.getStamper(stampWith);
            if (!activeStamper) {
                return undefined;
            }
            const { organizationId, timestampMs, ...parameters } = input;
            const session = await this.storageManager?.getActiveSession();
            const fullUrl = this.config.apiBaseUrl + "/public/v1/submit/update_private_key_tag";
            const bodyWithType = {
                parameters,
                organizationId: organizationId ?? session?.organizationId ?? this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_UPDATE_PRIVATE_KEY_TAG",
            };
            const stringifiedBody = JSON.stringify(bodyWithType);
            const stamp = await activeStamper.stamp(stringifiedBody);
            return {
                body: stringifiedBody,
                stamp: stamp,
                url: fullUrl,
            };
        };
        this.updateRootQuorum = async (input, stampWith) => {
            const { organizationId, timestampMs, ...rest } = input;
            const session = await this.storageManager?.getActiveSession();
            return this.activity("/public/v1/submit/update_root_quorum", {
                parameters: rest,
                organizationId: organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_UPDATE_ROOT_QUORUM",
            }, "updateRootQuorumResult", stampWith);
        };
        this.stampUpdateRootQuorum = async (input, stampWith) => {
            const activeStamper = this.getStamper(stampWith);
            if (!activeStamper) {
                return undefined;
            }
            const { organizationId, timestampMs, ...parameters } = input;
            const session = await this.storageManager?.getActiveSession();
            const fullUrl = this.config.apiBaseUrl + "/public/v1/submit/update_root_quorum";
            const bodyWithType = {
                parameters,
                organizationId: organizationId ?? session?.organizationId ?? this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_UPDATE_ROOT_QUORUM",
            };
            const stringifiedBody = JSON.stringify(bodyWithType);
            const stamp = await activeStamper.stamp(stringifiedBody);
            return {
                body: stringifiedBody,
                stamp: stamp,
                url: fullUrl,
            };
        };
        this.updateUser = async (input, stampWith) => {
            const { organizationId, timestampMs, ...rest } = input;
            const session = await this.storageManager?.getActiveSession();
            return this.activity("/public/v1/submit/update_user", {
                parameters: rest,
                organizationId: organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_UPDATE_USER",
            }, "updateUserResult", stampWith);
        };
        this.stampUpdateUser = async (input, stampWith) => {
            const activeStamper = this.getStamper(stampWith);
            if (!activeStamper) {
                return undefined;
            }
            const { organizationId, timestampMs, ...parameters } = input;
            const session = await this.storageManager?.getActiveSession();
            const fullUrl = this.config.apiBaseUrl + "/public/v1/submit/update_user";
            const bodyWithType = {
                parameters,
                organizationId: organizationId ?? session?.organizationId ?? this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_UPDATE_USER",
            };
            const stringifiedBody = JSON.stringify(bodyWithType);
            const stamp = await activeStamper.stamp(stringifiedBody);
            return {
                body: stringifiedBody,
                stamp: stamp,
                url: fullUrl,
            };
        };
        this.updateUserEmail = async (input, stampWith) => {
            const { organizationId, timestampMs, ...rest } = input;
            const session = await this.storageManager?.getActiveSession();
            return this.activity("/public/v1/submit/update_user_email", {
                parameters: rest,
                organizationId: organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_UPDATE_USER_EMAIL",
            }, "updateUserEmailResult", stampWith);
        };
        this.stampUpdateUserEmail = async (input, stampWith) => {
            const activeStamper = this.getStamper(stampWith);
            if (!activeStamper) {
                return undefined;
            }
            const { organizationId, timestampMs, ...parameters } = input;
            const session = await this.storageManager?.getActiveSession();
            const fullUrl = this.config.apiBaseUrl + "/public/v1/submit/update_user_email";
            const bodyWithType = {
                parameters,
                organizationId: organizationId ?? session?.organizationId ?? this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_UPDATE_USER_EMAIL",
            };
            const stringifiedBody = JSON.stringify(bodyWithType);
            const stamp = await activeStamper.stamp(stringifiedBody);
            return {
                body: stringifiedBody,
                stamp: stamp,
                url: fullUrl,
            };
        };
        this.updateUserName = async (input, stampWith) => {
            const { organizationId, timestampMs, ...rest } = input;
            const session = await this.storageManager?.getActiveSession();
            return this.activity("/public/v1/submit/update_user_name", {
                parameters: rest,
                organizationId: organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_UPDATE_USER_NAME",
            }, "updateUserNameResult", stampWith);
        };
        this.stampUpdateUserName = async (input, stampWith) => {
            const activeStamper = this.getStamper(stampWith);
            if (!activeStamper) {
                return undefined;
            }
            const { organizationId, timestampMs, ...parameters } = input;
            const session = await this.storageManager?.getActiveSession();
            const fullUrl = this.config.apiBaseUrl + "/public/v1/submit/update_user_name";
            const bodyWithType = {
                parameters,
                organizationId: organizationId ?? session?.organizationId ?? this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_UPDATE_USER_NAME",
            };
            const stringifiedBody = JSON.stringify(bodyWithType);
            const stamp = await activeStamper.stamp(stringifiedBody);
            return {
                body: stringifiedBody,
                stamp: stamp,
                url: fullUrl,
            };
        };
        this.updateUserPhoneNumber = async (input, stampWith) => {
            const { organizationId, timestampMs, ...rest } = input;
            const session = await this.storageManager?.getActiveSession();
            return this.activity("/public/v1/submit/update_user_phone_number", {
                parameters: rest,
                organizationId: organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_UPDATE_USER_PHONE_NUMBER",
            }, "updateUserPhoneNumberResult", stampWith);
        };
        this.stampUpdateUserPhoneNumber = async (input, stampWith) => {
            const activeStamper = this.getStamper(stampWith);
            if (!activeStamper) {
                return undefined;
            }
            const { organizationId, timestampMs, ...parameters } = input;
            const session = await this.storageManager?.getActiveSession();
            const fullUrl = this.config.apiBaseUrl + "/public/v1/submit/update_user_phone_number";
            const bodyWithType = {
                parameters,
                organizationId: organizationId ?? session?.organizationId ?? this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_UPDATE_USER_PHONE_NUMBER",
            };
            const stringifiedBody = JSON.stringify(bodyWithType);
            const stamp = await activeStamper.stamp(stringifiedBody);
            return {
                body: stringifiedBody,
                stamp: stamp,
                url: fullUrl,
            };
        };
        this.updateUserTag = async (input, stampWith) => {
            const { organizationId, timestampMs, ...rest } = input;
            const session = await this.storageManager?.getActiveSession();
            return this.activity("/public/v1/submit/update_user_tag", {
                parameters: rest,
                organizationId: organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_UPDATE_USER_TAG",
            }, "updateUserTagResult", stampWith);
        };
        this.stampUpdateUserTag = async (input, stampWith) => {
            const activeStamper = this.getStamper(stampWith);
            if (!activeStamper) {
                return undefined;
            }
            const { organizationId, timestampMs, ...parameters } = input;
            const session = await this.storageManager?.getActiveSession();
            const fullUrl = this.config.apiBaseUrl + "/public/v1/submit/update_user_tag";
            const bodyWithType = {
                parameters,
                organizationId: organizationId ?? session?.organizationId ?? this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_UPDATE_USER_TAG",
            };
            const stringifiedBody = JSON.stringify(bodyWithType);
            const stamp = await activeStamper.stamp(stringifiedBody);
            return {
                body: stringifiedBody,
                stamp: stamp,
                url: fullUrl,
            };
        };
        this.updateWallet = async (input, stampWith) => {
            const { organizationId, timestampMs, ...rest } = input;
            const session = await this.storageManager?.getActiveSession();
            return this.activity("/public/v1/submit/update_wallet", {
                parameters: rest,
                organizationId: organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_UPDATE_WALLET",
            }, "updateWalletResult", stampWith);
        };
        this.stampUpdateWallet = async (input, stampWith) => {
            const activeStamper = this.getStamper(stampWith);
            if (!activeStamper) {
                return undefined;
            }
            const { organizationId, timestampMs, ...parameters } = input;
            const session = await this.storageManager?.getActiveSession();
            const fullUrl = this.config.apiBaseUrl + "/public/v1/submit/update_wallet";
            const bodyWithType = {
                parameters,
                organizationId: organizationId ?? session?.organizationId ?? this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_UPDATE_WALLET",
            };
            const stringifiedBody = JSON.stringify(bodyWithType);
            const stamp = await activeStamper.stamp(stringifiedBody);
            return {
                body: stringifiedBody,
                stamp: stamp,
                url: fullUrl,
            };
        };
        this.verifyOtp = async (input, stampWith) => {
            const { organizationId, timestampMs, ...rest } = input;
            const session = await this.storageManager?.getActiveSession();
            return this.activity("/public/v1/submit/verify_otp", {
                parameters: rest,
                organizationId: organizationId ??
                    session?.organizationId ??
                    this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_VERIFY_OTP",
            }, "verifyOtpResult", stampWith);
        };
        this.stampVerifyOtp = async (input, stampWith) => {
            const activeStamper = this.getStamper(stampWith);
            if (!activeStamper) {
                return undefined;
            }
            const { organizationId, timestampMs, ...parameters } = input;
            const session = await this.storageManager?.getActiveSession();
            const fullUrl = this.config.apiBaseUrl + "/public/v1/submit/verify_otp";
            const bodyWithType = {
                parameters,
                organizationId: organizationId ?? session?.organizationId ?? this.config.organizationId,
                timestampMs: timestampMs ?? String(Date.now()),
                type: "ACTIVITY_TYPE_VERIFY_OTP",
            };
            const stringifiedBody = JSON.stringify(bodyWithType);
            const stamp = await activeStamper.stamp(stringifiedBody);
            return {
                body: stringifiedBody,
                stamp: stamp,
                url: fullUrl,
            };
        };
        this.proxyGetAccount = async (input) => {
            return this.authProxyRequest("/v1/account", input);
        };
        this.proxyOAuth2Authenticate = async (input) => {
            return this.authProxyRequest("/v1/oauth2_authenticate", input);
        };
        this.proxyOAuthLogin = async (input) => {
            return this.authProxyRequest("/v1/oauth_login", input);
        };
        this.proxyInitOtp = async (input) => {
            return this.authProxyRequest("/v1/otp_init", input);
        };
        this.proxyOtpLogin = async (input) => {
            return this.authProxyRequest("/v1/otp_login", input);
        };
        this.proxyVerifyOtp = async (input) => {
            return this.authProxyRequest("/v1/otp_verify", input);
        };
        this.proxySignup = async (input) => {
            return this.authProxyRequest("/v1/signup", input);
        };
        this.proxyGetWalletKitConfig = async (input) => {
            return this.authProxyRequest("/v1/wallet_kit_config", input);
        };
        this.config = config;
        if (config.apiKeyStamper) {
            this.apiKeyStamper = config.apiKeyStamper;
        }
        if (config.passkeyStamper) {
            this.passkeyStamper = config.passkeyStamper;
        }
        if (config.walletStamper) {
            this.walletStamper = config.walletStamper;
        }
        if (config.storageManager) {
            this.storageManager = config.storageManager;
        }
        if (config.defaultStamperType) {
            this.defaultStamperType = config.defaultStamperType;
        }
        else {
            // Set default stamper type based on available stampers
            if (this.apiKeyStamper) {
                this.defaultStamperType = StamperType.ApiKey;
            }
            else if (this.passkeyStamper) {
                this.defaultStamperType = StamperType.Passkey;
            }
            else if (this.walletStamper) {
                this.defaultStamperType = StamperType.Wallet;
            }
            else {
                this.defaultStamperType = undefined;
            }
        }
    }
    getStamper(stampWith) {
        if (!stampWith) {
            // Use default stamper type if none specified
            stampWith = this.defaultStamperType;
        }
        switch (stampWith) {
            case StamperType.ApiKey:
                return this.apiKeyStamper;
            case StamperType.Passkey:
                return this.passkeyStamper;
            case StamperType.Wallet:
                return this.walletStamper;
            default:
                return this.apiKeyStamper;
        }
    }
    sleep(ms) {
        return new Promise((resolve) => setTimeout(resolve, ms));
    }
    /**
     * Poll for activity completion until terminal status or max retries.
     * @internal
     */
    async pollForCompletion(activityId, stampWith) {
        const pollingDuration = this.config.activityPoller?.intervalMs ?? 1000;
        const maxRetries = this.config.activityPoller?.numRetries ?? 3;
        let attempts = 0;
        let activityData;
        do {
            await this.sleep(pollingDuration);
            activityData = (await this.getActivity({ activityId }, stampWith));
            attempts++;
        } while (!TERMINAL_ACTIVITY_STATUSES.includes(activityData.activity.status) &&
            attempts < maxRetries);
        return activityData;
    }
    /**
     * Extract and flatten result fields from a completed activity response.
     * @internal
     */
    handleActivityResponse(activityData, resultKey) {
        const { result, status } = activityData.activity;
        if (status === "ACTIVITY_STATUS_COMPLETED" && result) {
            // If a specific resultKey was provided, use it
            if (resultKey && result[resultKey]) {
                return {
                    ...result[resultKey],
                    ...activityData,
                };
            }
            // Otherwise, try to find any result field and flatten it
            for (const key of Object.keys(result)) {
                if (key.endsWith("Result") && result[key]) {
                    return {
                        ...result[key],
                        ...activityData,
                    };
                }
            }
        }
        return activityData;
    }
    async request(url, body, stampWith) {
        const fullUrl = this.config.apiBaseUrl + url;
        const stringifiedBody = JSON.stringify(body);
        var headers = {
            "Content-Type": "application/json",
            "X-Client-Version": VERSION,
        };
        // Use the specified stamper for this request
        const activeStamper = this.getStamper(stampWith);
        if (activeStamper) {
            const stamp = await activeStamper.stamp(stringifiedBody);
            headers[stamp.stampHeaderName] = stamp.stampHeaderValue;
        }
        const response = await fetch(fullUrl, {
            method: "POST",
            headers: headers,
            body: stringifiedBody,
            redirect: "follow",
        });
        if (!response.ok) {
            let res;
            try {
                res = await response.json();
            }
            catch (_) {
                throw new Error(`${response.status} ${response.statusText}`);
            }
            throw new TurnkeyRequestError(res);
        }
        const data = await response.json();
        return data;
    }
    async activity(url, body, resultKey, stampWith) {
        // Make the initial request
        let activityData = await this.request(url, body, stampWith);
        // Poll if not in terminal status
        if (!TERMINAL_ACTIVITY_STATUSES.includes(activityData.activity.status)) {
            activityData = await this.pollForCompletion(activityData.activity.id, stampWith);
        }
        return this.handleActivityResponse(activityData, resultKey);
    }
    async activityDecision(url, body, stampWith) {
        // Use the specified stamper for this request
        const activityData = (await this.request(url, body, stampWith));
        return {
            ...activityData["activity"]["result"],
            ...activityData,
        };
    }
    async authProxyRequest(url, body) {
        if (!this.config.authProxyUrl || !this.config.authProxyConfigId) {
            throw new TurnkeyError("Auth Proxy URL or ID is not configured.", TurnkeyErrorCodes.INVALID_CONFIGURATION);
        }
        const fullUrl = this.config.authProxyUrl + url;
        const stringifiedBody = JSON.stringify(body);
        var headers = {
            "Content-Type": "application/json",
            "X-Auth-Proxy-Config-ID": this.config.authProxyConfigId,
        };
        const response = await fetch(fullUrl, {
            method: "POST",
            headers: headers,
            body: stringifiedBody,
        });
        if (!response.ok) {
            let res;
            try {
                res = await response.json();
            }
            catch (_) {
                throw new Error(`${response.status} ${response.statusText}`);
            }
            throw new TurnkeyRequestError(res);
        }
        const data = await response.json();
        return data;
    }
    /**
     * Submit a pre-signed request to Turnkey.
     *
     * Use this method to execute any TSignedRequest returned by the SDK's
     * stamping methods (stampCreateApiKeys, stampGetPolicies, stampGetWallets, etc.).
     *
     * Works for both query and activity requests:
     * - For queries: returns the response directly
     * - For activities: automatically polls until completion and extracts the result
     *
     * @param signedRequest - A TSignedRequest object returned by a stamping method
     * @param options - Optional configuration for the request
     * @param options.resultKey - For activity requests, the key to extract from the result (e.g., "createApiKeysResultV2")
     * @returns The parsed response, with activity result fields flattened if applicable
     */
    async sendSignedRequest(signedRequest, options) {
        const headers = {
            "Content-Type": "application/json",
            "X-Client-Version": VERSION,
            [signedRequest.stamp.stampHeaderName]: signedRequest.stamp.stampHeaderValue,
        };
        const response = await fetch(signedRequest.url, {
            method: "POST",
            headers,
            body: signedRequest.body,
        });
        if (!response.ok) {
            let res;
            try {
                res = await response.json();
            }
            catch (_) {
                throw new Error(`${response.status} ${response.statusText}`);
            }
            throw new TurnkeyRequestError(res);
        }
        const data = await response.json();
        // Check if this is an activity response that needs polling
        if (data.activity) {
            let activityData = data;
            // Poll if not in terminal status
            if (!TERMINAL_ACTIVITY_STATUSES.includes(activityData.activity.status)) {
                activityData = await this.pollForCompletion(activityData.activity.id);
            }
            return this.handleActivityResponse(activityData, options?.resultKey);
        }
        return data;
    }
}

export { TurnkeySDKClientBase };
