// @turnkey/core/dist/__stampers__/api/base.js
'use strict';

var utils = require('../../utils.js');
var stamper = require('./web/stamper.js');
var sdkTypes = require('@turnkey/sdk-types');
var apiKeyStamper = require('@turnkey/api-key-stamper');

/**
 * Cross-platform API key stamper.
 *
 * - This stamper uses indexedDB on web and keychain on react-native to securely stamp Turnkey requests.
 * - ***Only supports P-256 ECDSA key pairs.***
 */
class CrossPlatformApiKeyStamper {
    constructor(storageManager) {
        this.storageManager = storageManager;
        // Use init method to set up the stamper based on the platform. It's async, so can't be done in the constructor.
    }
    async init() {
        if (utils.isWeb()) {
            this.stamper = new stamper.IndexedDbStamper();
        }
        else if (utils.isReactNative()) {
            try {
                // Dynamic import to prevent bundling the native module in web environments.
                const { ReactNativeKeychainStamper } = await Promise.resolve().then(function () { return require('./mobile/stamper.js'); });
                this.stamper = new ReactNativeKeychainStamper();
            }
            catch (error) {
                throw new sdkTypes.TurnkeyError(`Failed to load keychain stamper for react-native`, sdkTypes.TurnkeyErrorCodes.INITIALIZE_CLIENT_ERROR, error);
            }
        }
        else {
            throw new sdkTypes.TurnkeyError("Unsupported platform for API key stamper", sdkTypes.TurnkeyErrorCodes.UNSUPPORTED_PLATFORM);
        }
    }
    listKeyPairs() {
        if (!this.stamper) {
            throw new sdkTypes.TurnkeyError("Stamper is not initialized. Please call .init() before calling this method.", sdkTypes.TurnkeyErrorCodes.CLIENT_NOT_INITIALIZED);
        }
        return this.stamper.listKeyPairs();
    }
    createKeyPair(externalKeyPair) {
        if (!this.stamper) {
            throw new sdkTypes.TurnkeyError("Stamper is not initialized. Please call .init() before calling this method.", sdkTypes.TurnkeyErrorCodes.CLIENT_NOT_INITIALIZED);
        }
        return this.stamper.createKeyPair(externalKeyPair);
    }
    deleteKeyPair(publicKeyHex) {
        if (!this.stamper) {
            throw new sdkTypes.TurnkeyError("Stamper is not initialized. Please call .init() before calling this method.", sdkTypes.TurnkeyErrorCodes.CLIENT_NOT_INITIALIZED);
        }
        // If the deleted key pair is the temporary one, clear it.
        if (this.temporaryPublicKey === publicKeyHex) {
            this.temporaryPublicKey = undefined;
        }
        return this.stamper.deleteKeyPair(publicKeyHex);
    }
    // This allows forcing a specific public key to find the key pair for stamping. The key pair must already exist in indexedDB / Keychain.
    // This is useful if you need to stamp with a specific key pair without having an active session.
    // See "signUpWithPasskey" function in core.ts for usage
    setTemporaryPublicKey(publicKeyHex) {
        this.temporaryPublicKey = publicKeyHex;
    }
    getTemporaryPublicKey() {
        return this.temporaryPublicKey;
    }
    clearTemporaryPublicKey() {
        this.temporaryPublicKey = undefined;
    }
    async stamp(payload) {
        if (!this.stamper) {
            throw new sdkTypes.TurnkeyError("Stamper is not initialized. Please call .init() before calling this method.", sdkTypes.TurnkeyErrorCodes.CLIENT_NOT_INITIALIZED);
        }
        let publicKeyHex = this.temporaryPublicKey;
        if (!publicKeyHex) {
            const session = await this.storageManager.getActiveSession();
            if (!session) {
                throw new sdkTypes.TurnkeyError("No active session or token available.", sdkTypes.TurnkeyErrorCodes.NO_SESSION_FOUND);
            }
            publicKeyHex = session.publicKey;
        }
        return this.stamper.stamp(payload, publicKeyHex);
    }
    async sign(payload, format = apiKeyStamper.SignatureFormat.Der) {
        if (!this.stamper) {
            throw new sdkTypes.TurnkeyError("Stamper is not initialized. Please call .init() before calling this method.", sdkTypes.TurnkeyErrorCodes.CLIENT_NOT_INITIALIZED);
        }
        let publicKeyHex = this.temporaryPublicKey;
        if (!publicKeyHex) {
            const session = await this.storageManager.getActiveSession();
            if (!session) {
                throw new sdkTypes.TurnkeyError("No active session or token available.", sdkTypes.TurnkeyErrorCodes.NO_SESSION_FOUND);
            }
            publicKeyHex = session.publicKey;
        }
        return this.stamper.sign(payload, publicKeyHex, format);
    }
}

exports.CrossPlatformApiKeyStamper = CrossPlatformApiKeyStamper;
