// @turnkey/core/dist/index.js
'use strict';

var apiKeyStamper = require('@turnkey/api-key-stamper');
var webauthnStamper = require('@turnkey/webauthn-stamper');
var core = require('./__clients__/core.js');
var auth = require('./__types__/auth.js');
var enums = require('./__types__/enums.js');
var error = require('./__types__/error.js');
var http = require('./__types__/http.js');
var sdkTypes = require('@turnkey/sdk-types');
var utils = require('./utils.js');



Object.defineProperty(exports, "ApiKeyStamper", {
	enumerable: true,
	get: function () { return apiKeyStamper.ApiKeyStamper; }
});
Object.defineProperty(exports, "signWithApiKey", {
	enumerable: true,
	get: function () { return apiKeyStamper.signWithApiKey; }
});
Object.defineProperty(exports, "WebauthnStamper", {
	enumerable: true,
	get: function () { return webauthnStamper.WebauthnStamper; }
});
exports.TurnkeyClient = core.TurnkeyClient;
exports.DEFAULT_SESSION_EXPIRATION_IN_SECONDS = auth.DEFAULT_SESSION_EXPIRATION_IN_SECONDS;
Object.defineProperty(exports, "Chain", {
	enumerable: true,
	get: function () { return enums.Chain; }
});
Object.defineProperty(exports, "Curve", {
	enumerable: true,
	get: function () { return enums.Curve; }
});
Object.defineProperty(exports, "FilterType", {
	enumerable: true,
	get: function () { return enums.FilterType; }
});
Object.defineProperty(exports, "OtpType", {
	enumerable: true,
	get: function () { return enums.OtpType; }
});
exports.OtpTypeToFilterTypeMap = enums.OtpTypeToFilterTypeMap;
Object.defineProperty(exports, "SessionKey", {
	enumerable: true,
	get: function () { return enums.SessionKey; }
});
Object.defineProperty(exports, "SignIntent", {
	enumerable: true,
	get: function () { return enums.SignIntent; }
});
Object.defineProperty(exports, "StamperType", {
	enumerable: true,
	get: function () { return enums.StamperType; }
});
Object.defineProperty(exports, "WalletInterfaceType", {
	enumerable: true,
	get: function () { return enums.WalletInterfaceType; }
});
Object.defineProperty(exports, "WalletSource", {
	enumerable: true,
	get: function () { return enums.WalletSource; }
});
exports.TurnkeyRequestError = error.TurnkeyRequestError;
exports.TERMINAL_ACTIVITY_STATUSES = http.TERMINAL_ACTIVITY_STATUSES;
exports.addressFormatConfig = utils.addressFormatConfig;
exports.decodeVerificationToken = utils.decodeVerificationToken;
exports.generateWalletAccountsFromAddressFormat = utils.generateWalletAccountsFromAddressFormat;
exports.getAuthProxyConfig = utils.getAuthProxyConfig;
exports.getClientSignatureMessageForLogin = utils.getClientSignatureMessageForLogin;
exports.getClientSignatureMessageForSignup = utils.getClientSignatureMessageForSignup;
exports.isEthereumProvider = utils.isEthereumProvider;
exports.isSolanaProvider = utils.isSolanaProvider;
exports.sendSignedRequest = utils.sendSignedRequest;
Object.keys(sdkTypes).forEach(function (k) {
	if (k !== 'default' && !Object.prototype.hasOwnProperty.call(exports, k)) Object.defineProperty(exports, k, {
		enumerable: true,
		get: function () { return sdkTypes[k]; }
	});
});
