// @turnkey/core/dist/__storage__/web/storage.js
'use strict';

var window = require('../../__polyfills__/window.js');
var enums = require('../../__types__/enums.js');
var utils = require('../../utils.js');
var sdkTypes = require('@turnkey/sdk-types');

const browserStorage = window.localStorage;
class WebStorageManager {
    constructor() {
        this.getStorageValue = async (sessionKey) => {
            const item = browserStorage.getItem(sessionKey);
            return item ? JSON.parse(item) : undefined;
        };
        this.setStorageValue = async (sessionKey, storageValue) => {
            if (storageValue === undefined) {
                throw new sdkTypes.TurnkeyError("Session value cannot be undefined", sdkTypes.TurnkeyErrorCodes.STORE_SESSION_ERROR);
            }
            browserStorage.setItem(sessionKey, JSON.stringify(storageValue));
        };
        this.setActiveSessionKey = async (sessionKey) => {
            await this.setStorageValue(WebStorageManager.ACTIVE_SESSION_KEY, sessionKey);
        };
        this.removeStorageValue = async (sessionKey) => {
            browserStorage.removeItem(sessionKey);
        };
        this.storeSession = async (session, sessionKey = enums.SessionKey.DefaultSessionkey) => {
            const sessionWithMetadata = utils.parseSession(session);
            await this.setStorageValue(sessionKey, sessionWithMetadata);
            // Ensure the session key is stored in the session keys list
            const keys = (await this.getStorageValue(WebStorageManager.ALL_SESSION_KEYS)) ?? [];
            if (!keys.includes(sessionKey)) {
                keys.push(sessionKey);
                await this.setStorageValue(WebStorageManager.ALL_SESSION_KEYS, keys);
            }
            // Set the active session key
            await this.setStorageValue(WebStorageManager.ACTIVE_SESSION_KEY, sessionKey);
        };
        this.getSession = async (sessionKey = enums.SessionKey.DefaultSessionkey) => {
            return this.getStorageValue(sessionKey);
        };
        this.getActiveSessionKey = async () => {
            return this.getStorageValue(WebStorageManager.ACTIVE_SESSION_KEY);
        };
        this.getActiveSession = async () => {
            const key = await this.getActiveSessionKey();
            return key ? this.getSession(key) : undefined;
        };
        this.listSessionKeys = async () => {
            return ((await this.getStorageValue(WebStorageManager.ALL_SESSION_KEYS)) ?? []);
        };
        this.clearSession = async (sessionKey) => {
            await this.removeStorageValue(sessionKey);
            const keys = await this.listSessionKeys();
            const updated = keys.filter((k) => k !== sessionKey);
            await this.setStorageValue(WebStorageManager.ALL_SESSION_KEYS, updated);
            const active = await this.getActiveSessionKey();
            if (active === sessionKey) {
                await this.removeStorageValue(WebStorageManager.ACTIVE_SESSION_KEY);
            }
        };
        this.clearAllSessions = async () => {
            const keys = await this.listSessionKeys();
            await Promise.all(keys.map((k) => this.removeStorageValue(k)));
            await this.removeStorageValue(WebStorageManager.ALL_SESSION_KEYS);
            await this.removeStorageValue(WebStorageManager.ACTIVE_SESSION_KEY);
        };
        this.storeWallets = async (wallets) => {
            for (const wallet of wallets) {
                browserStorage.setItem(wallet.walletId, JSON.stringify(wallet));
            }
        };
    }
}
WebStorageManager.ALL_SESSION_KEYS = "@turnkey/all-session-keys";
WebStorageManager.ACTIVE_SESSION_KEY = "@turnkey/active-session-key";

exports.WebStorageManager = WebStorageManager;
